{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b94817",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/scratch2/jhsiao/projects/notebooks/book-top-directory/metazoa16-mapped-repaired/1-mapping-pacbio')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1811ebe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#001-wget-genome-pacbio-transcriptome\n",
    "\n",
    "mkdir output || echo \"\"\n",
    "mkdir output/source-data/ || echo \"\"\n",
    "\n",
    "# PacBio transcriptome NCBI TSA\n",
    "wget http://igc2.salk.edu/illumina/runs/PacBio_200525/Chalasani_Eric_200525/EE04/ISO-seq3/hq_transcripts.fasta -P output/\n",
    "\n",
    "# Ensembl genome mRNA\n",
    "wget http://ftp.ensemblgenomes.org/pub/metazoa/release-52/fasta/mnemiopsis_leidyi/cdna/Mnemiopsis_leidyi.MneLei_Aug2011.cdna.all.fa.gz -P output/source-data/\n",
    "\n",
    "# Ensembl genome CDS\n",
    "wget http://ftp.ensemblgenomes.org/pub/metazoa/release-52/fasta/mnemiopsis_leidyi/cds/Mnemiopsis_leidyi.MneLei_Aug2011.cds.all.fa.gz -P output/source-data/\n",
    "\n",
    "# Ensembl genome AA\n",
    "wget http://ftp.ensemblgenomes.org/pub/metazoa/release-52/fasta/mnemiopsis_leidyi/pep/Mnemiopsis_leidyi.MneLei_Aug2011.pep.all.fa.gz -P output/source-data/\n",
    "\n",
    "# Ensembl genome GFF\n",
    "wget http://ftp.ensemblgenomes.org/pub/metazoa/release-52/gff3/mnemiopsis_leidyi/Mnemiopsis_leidyi.MneLei_Aug2011.52.gff3.gz -P output/source-data/ \n",
    "\n",
    "gunzip output/source-data/*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ebb78f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#002-evigene\n",
    "\n",
    "\n",
    "mkdir output/2-evigene || echo \"\"\n",
    "cd output/2-evigene\n",
    "cp ../source-data/hq_transcripts.fasta . &&\n",
    "\n",
    "trformat.pl  -output hq_transcripts-evigene_headers.fasta  -input hq_transcripts.fasta -format=pacbio &&\n",
    "tr2aacds4.pl -mrnaseq hq_transcripts-evigene_headers.fasta -NCPU=40 -MAXMEM=900000 -MINAA=30 -logfile 2-log-hq_transcripts -tidyup \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9f127f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#003-python-parse-evigene-T1s\n",
    "\n",
    "\n",
    "# >NonamEVm000019t1 type=mRNA; aalen=3767,77%,complete; clen=14673; offs=2676-13979; organism=No_name; evgclass=main; oid=evgpacbUnnamedSample_HQ_transcript/1;\n",
    "# >NonamEVm000019t1 type=CDS; aalen=3767,77%,complete; clen=14673; offs=2676-13979; organism=No_name; evgclass=main; oid=evgpacbUnnamedSample_HQ_transcript/1; codepot=Code/0.53;\n",
    "# >NonamEVm000019t1 type=protein; aalen=3767,77%,complete; clen=14673; offs=2676-13979; organism=No_name; evgclass=main; oid=evgpacbUnnamedSample_HQ_transcript/1; codepot=Code/0.53;\n",
    "\n",
    "output_map = open( 'output/3-map-pacbio-to-evigeneT1-headers', 'w' )\n",
    "\n",
    "file_types = [ 'mrna', 'cds', 'aa' ]\n",
    "\n",
    "for next_type in file_types:\n",
    "\n",
    "    file_in = 'output/2-evigene/okayset/hq_transcripts-evigene_headers.okay.' + next_type\n",
    "    input_fasta = open( file_in, 'r' )\n",
    "    \n",
    "    file_out = 'output/3-Mnemiopsis-leidyi-pacbio-evigeneT1.' + next_type\n",
    "    output_fasta = open( file_out, 'w' )\n",
    "\n",
    "    header_seq = {}\n",
    "    source_ids = []\n",
    "    for next_line in input_fasta:\n",
    "\n",
    "        if next_line[ 0 ] == '>':\n",
    "            info_1 = next_line[ 1:-1 ].split( ' ' )\n",
    "            info_2 = info_1[ 0 ].split( 't' )\n",
    "\n",
    "            if info_2[ -1 ] == '1':\n",
    "\n",
    "                new_id = 'Mnemiopsis-leidyi-' + info_1[ 0 ]\n",
    "\n",
    "                info_3 = next_line[ :-1 ].split( ' oid=' )\n",
    "                info_4 = info_3[ -1 ].split( ';' )\n",
    "                info_5 = info_4[ 0 ].split( '_' )\n",
    "                source_id = '_'.join( info_5[ 1: ] )\n",
    "                                \n",
    "                # this will drop additional utr isoforms of a sequence - evigene adds them as additional T1s\n",
    "                # but I think this is incorrect\n",
    "                # from the perspective of T1s should represent a single isoform of a gene in the genome\n",
    "                \n",
    "                if new_id in header_seq.keys():\n",
    "                    keeper = False\n",
    "\n",
    "                else:\n",
    "\n",
    "                    source_id_clean_list = source_id.split( 'utrorf' )\n",
    "                    source_id_clean = source_id_clean_list[ 0 ]\n",
    "                    \n",
    "                    if len( source_id_clean_list ) > 1:\n",
    "                        utr_issue = 'utrorf;' + info_4[ 1 ].split( ' ' )[ 0 ]\n",
    "\n",
    "                    else:\n",
    "                        utr_issue = ''\n",
    "                        \n",
    "                    header_seq[ new_id ] = ''\n",
    "                    keeper = True\n",
    "\n",
    "                    output = next_type + '\\t' + new_id + '\\t' + source_id_clean + '\\t' + utr_issue +  '\\n'\n",
    "                    output_map.write( output )\n",
    "           \n",
    "            else:\n",
    "                keeper = False\n",
    "\n",
    "        else:\n",
    "\n",
    "            if keeper == True:\n",
    "                header_seq[ new_id ] = header_seq[ new_id ] + next_line[ :-1 ]\n",
    "\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "    for next_id in header_seq.keys():\n",
    "\n",
    "        header = '>' + next_id \n",
    "        sequence = header_seq[ next_id ]\n",
    "        output = header + '\\n' + sequence + '\\n'\n",
    "        output_fasta.write( output )\n",
    "\n",
    "    input_fasta.close()\n",
    "    output_fasta.close()\n",
    "\n",
    "output_map.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24cdf1d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#004-mkblastdb-genome\n",
    "\n",
    "\n",
    "/usr/bin/makeblastdb -in output/Mnemiopsis_leidyi.MneLei_Aug2011.cdna.all.fa -parse_seqids -dbtype nucl\n",
    "/usr/bin/makeblastdb -in output/Mnemiopsis_leidyi.MneLei_Aug2011.cds.all.fa -parse_seqids -dbtype nucl\n",
    "/usr/bin/makeblastdb -in output/Mnemiopsis_leidyi.MneLei_Aug2011.pep.all.fa -parse_seqids -dbtype prot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a5de3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#005-blastn-pacbio_X_genome\n",
    "\n",
    "/usr/bin/blastn -db output/Mnemiopsis_leidyi.MneLei_Aug2011.cds.all.fa -outfmt 6 -out output/5-blast-report-pacbio_X_genome-Mnemiopsis-leidyi-CDS -max_hsps 1 -query output/3-Mnemiopsis-leidyi-pacbio-evigeneT1.mrna -evalue 1e-3 -num_threads 40\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c4b1eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#006-python-parse-report-and-update-gene-models-MRNA\n",
    "\n",
    "# UnnamedSample_HQ_transcript/0   ML16643a-RA     99.804  5090    10      0       9617    14706   1048    6137    0.0     9345\n",
    "\n",
    "input_report = open( 'output/5-blast-report-pacbio_X_genome-Mnemiopsis-leidyi-CDS', 'r' )\n",
    "input_transcriptome = open( 'output/3-Mnemiopsis-leidyi-pacbio-evigeneT1.mrna', 'r' )\n",
    "input_genome = open( 'output/Mnemiopsis_leidyi.MneLei_Aug2011.cdna.all.fa', 'r' )\n",
    "\n",
    "output_fasta = open( 'output/6-Mnemiopsis_leidyi-genome-pacbio-updates.mrna', 'w' )\n",
    "output_map = open( 'output/6-map-sourceid-to-updateid', 'w' )\n",
    "output_difference = open( 'output/6-pacbio-minus-gene-lengths', 'w' )\n",
    "output_stats = open( 'output/6-stats', 'w' )\n",
    "output_report_stats = open( 'output/6-stats-more-than-one-gene-at-0.0', 'w' )\n",
    "\n",
    "# read in genome sequences\n",
    "genome_seq = {}\n",
    "transcriptome_seq = {}\n",
    "genome_original_letters = ''\n",
    "for next_line in input_genome:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "        header_id = next_line[ 1:-1 ].split( ' ' )[ 0 ]\n",
    "        genome_seq[ header_id ] = ''\n",
    "\n",
    "    else:\n",
    "        sequence = next_line[ :-1 ]\n",
    "        genome_seq[ header_id ] = genome_seq[ header_id ] + sequence\n",
    "        genome_original_letters = genome_original_letters +  sequence\n",
    "        \n",
    "# read in pacbio transcriptome sequences\n",
    "for next_line in input_transcriptome:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "        header = next_line[ 1:-1 ]\n",
    "        transcriptome_seq[ header ] = ''\n",
    "\n",
    "    else:\n",
    "        transcriptome_seq[ header ] = transcriptome_seq[ header ] + next_line[ :-1 ]\n",
    "\n",
    "# 1.  qseqid      query or source (e.g., gene) sequence id\n",
    "# 2.  sseqid      subject  or target (e.g., reference genome) sequence id\n",
    "# 3.  pident      percentage of identical matches\n",
    "# 4.  length      alignment length (sequence overlap)\n",
    "# 5.  mismatch    number of mismatches\n",
    "# 6.  gapopen     number of gap openings\n",
    "# 7.  qstart      start of alignment in query\n",
    "# 8.  qend        end of alignment in query\n",
    "# 9.  sstart      start of alignment in subject\n",
    "# 10.  send        end of alignment in subject\n",
    "# 11.  evalue      expect value\n",
    "# 12.  bitscore    bit score\n",
    "\n",
    "# read in blast report\n",
    "transcript_genes = {}\n",
    "gene_transcripts = {}\n",
    "transcript_report = {}\n",
    "for next_line in input_report:\n",
    "\n",
    "    info = next_line.split( '\\t' )\n",
    "    transcript_id = info[ 0 ]\n",
    "    gene_id = info[ 1 ]\n",
    "    evalue = info[ 10 ]\n",
    "    percent_id = info[ 2 ]\n",
    "    align_length = info[ 3 ]\n",
    "    mismatch = info[ 4 ]\n",
    "    bitscore = info[ 11 ]\n",
    "\n",
    "    transcript_stats = [ transcript_id, gene_id, evalue, percent_id, align_length, mismatch, bitscore ]\n",
    "    \n",
    "    if gene_id in genome_seq.keys():\n",
    "        \n",
    "        if evalue == '0.0':\n",
    " \n",
    "            # add transcript and genes to transcript_genes dictionary \n",
    "            if transcript_id in transcript_genes.keys():\n",
    "                transcript_genes[ transcript_id ].append( gene_id )\n",
    "                transcript_report[ transcript_id ][ gene_id ] = transcript_stats\n",
    "\n",
    "            else:\n",
    "                transcript_genes[ transcript_id ] = []\n",
    "                transcript_genes[ transcript_id ].append( gene_id )\n",
    "\n",
    "                transcript_report[ transcript_id ] = {}\n",
    "                transcript_report[ transcript_id ][ gene_id ] = transcript_stats\n",
    "                \n",
    "            # add gene and transcript to gene_transcripts dictionary\n",
    "            if gene_id in gene_transcripts.keys():\n",
    "                gene_transcripts[ gene_id ].append( transcript_id )\n",
    "\n",
    "            else:\n",
    "                gene_transcripts[ gene_id ] = []\n",
    "                gene_transcripts[ gene_id ].append( transcript_id )\n",
    "                                            \n",
    "        else:\n",
    "            pass\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "# parse for genes with just one hit have evalue 0.0\n",
    "gene_replacement = {}\n",
    "for next_transcript in transcript_genes.keys():\n",
    "\n",
    "    if len( transcript_genes[ next_transcript ] ) > 1:\n",
    "            \n",
    "        for next_gene in transcript_genes[ next_transcript ]:\n",
    "            gene_report_stats = transcript_report[ next_transcript ][ next_gene ]\n",
    "            output = ''\n",
    "            for next_stat in gene_report_stats:\n",
    "                output = output + next_stat + '\\t'\n",
    "            output = output[ :-1 ] + '\\n'\n",
    "            output_report_stats.write( output )\n",
    "\n",
    "    else:\n",
    "        gene_id = transcript_genes[ next_transcript ][ 0 ]\n",
    "        gene_sequence = genome_seq[ gene_id ]\n",
    "        transcript_sequence = transcriptome_seq[ next_transcript ]\n",
    "\n",
    "        gene_base = gene_id.split( '-' )[ 0 ]\n",
    "\n",
    "        if gene_base in gene_replacement.keys():\n",
    "            if len( transcriptome_seq[ next_transcript ] ) > len( gene_replacement[ gene_base ] ):\n",
    "                gene_replacement[ gene_base ] = next_transcript\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            gene_replacement[ gene_base ] = next_transcript\n",
    "\n",
    "# update genome gene model if there is a  pacbio sequence - assume pacbio sequence is better since read is real - gene model is hypothesized based on evidence of varying unknown specific quality\n",
    "# >ML49653a-RA cdna supercontig:MneLei_Aug2011:ML4965:35353:38358:-1 gene:ML49653a gene_biotype:protein_coding transcript_biotype:protein_coding_pacbioMnemiopsis-leidyi-NonamEVm004368t1\n",
    "\n",
    "genome_updated_letters = ''\n",
    "for next_gene in genome_seq.keys():\n",
    "    next_gene_base = next_gene.split( '-' )[ 0 ]\n",
    "    \n",
    "    if next_gene_base in gene_replacement.keys():\n",
    "        transcript_id = gene_replacement[ next_gene_base ]\n",
    "        next_gene_info = next_gene.split( ' ' )\n",
    "        transcript_info_1 = transcript_id.split( ' ' )\n",
    "        transcript_info_2 = transcript_info_1[ 0 ].split( '-' )\n",
    "        gene_id = next_gene_info[ 0 ]\n",
    "        original_seq = genome_seq[ gene_id ]\n",
    "        updated_gene_id = gene_id + '___pacbio' + transcript_info_2[ -1 ] #+ ' ' + ' '.join( next_gene_info[ 1: ] )\n",
    "        updated_seq = transcriptome_seq[ transcript_id ]\n",
    "\n",
    "        difference = str( len( updated_seq ) - len( original_seq ) )\n",
    "        output = 'PacBio ' + transcript_id + ' ' + str( len( updated_seq ) ) + '\\t Gene ' + gene_id + ' ' + str( len( original_seq ) ) + '\\t PacBio - Gene = ' + difference + ' nts mRNA\\n'\n",
    "        output_stats.write( output )\n",
    "        output = str( len( updated_seq ) ) + '\\t' + str( len( original_seq ) ) + '\\t' + difference + '\\n'\n",
    "        output_difference.write( output)\n",
    "\n",
    "        output = '>' + updated_gene_id + '\\n' + updated_seq + '\\n'\n",
    "        output_fasta.write( output )\n",
    "        genome_updated_letters = genome_updated_letters + updated_seq\n",
    "\n",
    "        output = next_gene + '\\t' + transcript_id + '\\t' + updated_gene_id + '\\n'\n",
    "        output_map.write( output )\n",
    "        \n",
    "    else:\n",
    "        sequence = genome_seq[ next_gene ]\n",
    "        output = '>' + next_gene + '\\n' + sequence + '\\n'\n",
    "        output_fasta.write( output )\n",
    "        genome_updated_letters = genome_updated_letters + sequence\n",
    "\n",
    "output = 'Original gene models = ' + str( len( genome_original_letters ) ) + ' nucleotide mRNA sequence'\n",
    "print( output)\n",
    "output_stats.write( output )\n",
    "output = 'Updated gene models = ' + str( len( genome_updated_letters ) ) + ' nucleotide mRNA sequence'\n",
    "print( output)\n",
    "output_stats.write( output )\n",
    "output = 'Number of nucleotides added to mRNA gene models = ' + str( len( genome_updated_letters ) - len( genome_original_letters ) )\n",
    "print( output )\n",
    "output_stats.write( output )\n",
    "\n",
    "input_report.close()\n",
    "input_transcriptome.close()\n",
    "input_genome.close()\n",
    "output_fasta.close()\n",
    "output_map.close()\n",
    "output_difference.close()\n",
    "output_stats.close()\n",
    "output_report_stats.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d859f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#007-python-parse-report-and-update-gene-models-CDS\n",
    "\n",
    "# UnnamedSample_HQ_transcript/0   ML16643a-RA     99.804  5090    10      0       9617    14706   1048    6137    0.0     9345\n",
    "\n",
    "input_report = open( 'output/5-blast-report-pacbio_X_genome-Mnemiopsis-leidyi-CDS', 'r' )\n",
    "input_transcriptome = open( 'output/3-Mnemiopsis-leidyi-pacbio-evigeneT1.cds', 'r' )\n",
    "input_genome = open( 'output/Mnemiopsis_leidyi.MneLei_Aug2011.cds.all.fa', 'r' )\n",
    "\n",
    "output_fasta = open( 'output/7-Mnemiopsis_leidyi-genome-pacbio-updates.cds', 'w' )\n",
    "output_map = open( 'output/7-map-sourceid-to-updateid', 'w' )\n",
    "output_difference = open( 'output/7-pacbio-minus-gene-lengths', 'w' )\n",
    "output_stats = open( 'output/7-stats', 'w' )\n",
    "output_report_stats = open( 'output/7-stats-more-than-one-gene-at-0.0', 'w' )\n",
    "\n",
    "# read in genome sequences\n",
    "genome_seq = {}\n",
    "transcriptome_seq = {}\n",
    "genome_original_letters = ''\n",
    "for next_line in input_genome:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "        header_id = next_line[ 1:-1 ].split( ' ' )[ 0 ]\n",
    "        genome_seq[ header_id ] = ''\n",
    "\n",
    "    else:\n",
    "        sequence = next_line[ :-1 ]\n",
    "        genome_seq[ header_id ] = genome_seq[ header_id ] + sequence\n",
    "        genome_original_letters = genome_original_letters +  sequence\n",
    "        \n",
    "# read in pacbio transcriptome sequences\n",
    "for next_line in input_transcriptome:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "        header = next_line[ 1:-1 ]\n",
    "        transcriptome_seq[ header ] = ''\n",
    "\n",
    "    else:\n",
    "        transcriptome_seq[ header ] = transcriptome_seq[ header ] + next_line[ :-1 ]\n",
    "\n",
    "# 1.  qseqid      query or source (e.g., gene) sequence id\n",
    "# 2.  sseqid      subject  or target (e.g., reference genome) sequence id\n",
    "# 3.  pident      percentage of identical matches\n",
    "# 4.  length      alignment length (sequence overlap)\n",
    "# 5.  mismatch    number of mismatches\n",
    "# 6.  gapopen     number of gap openings\n",
    "# 7.  qstart      start of alignment in query\n",
    "# 8.  qend        end of alignment in query\n",
    "# 9.  sstart      start of alignment in subject\n",
    "# 10.  send        end of alignment in subject\n",
    "# 11.  evalue      expect value\n",
    "# 12.  bitscore    bit score\n",
    "\n",
    "# read in blast report\n",
    "transcript_genes = {}\n",
    "gene_transcripts = {}\n",
    "transcript_report = {}\n",
    "for next_line in input_report:\n",
    "\n",
    "    info = next_line.split( '\\t' )\n",
    "    transcript_id = info[ 0 ]\n",
    "    gene_id = info[ 1 ]\n",
    "    evalue = info[ 10 ]\n",
    "    percent_id = info[ 2 ]\n",
    "    align_length = info[ 3 ]\n",
    "    mismatch = info[ 4 ]\n",
    "    bitscore = info[ 11 ]\n",
    "\n",
    "    transcript_stats = [ transcript_id, gene_id, evalue, percent_id, align_length, mismatch, bitscore ]\n",
    "    \n",
    "    if gene_id in genome_seq.keys():\n",
    "        \n",
    "        if evalue == '0.0':\n",
    " \n",
    "            # add transcript and genes to transcript_genes dictionary \n",
    "            if transcript_id in transcript_genes.keys():\n",
    "                transcript_genes[ transcript_id ].append( gene_id )\n",
    "                transcript_report[ transcript_id ][ gene_id ] = transcript_stats\n",
    "\n",
    "            else:\n",
    "                transcript_genes[ transcript_id ] = []\n",
    "                transcript_genes[ transcript_id ].append( gene_id )\n",
    "\n",
    "                transcript_report[ transcript_id ] = {}\n",
    "                transcript_report[ transcript_id ][ gene_id ] = transcript_stats\n",
    "                \n",
    "            # add gene and transcript to gene_transcripts dictionary\n",
    "            if gene_id in gene_transcripts.keys():\n",
    "                gene_transcripts[ gene_id ].append( transcript_id )\n",
    "\n",
    "            else:\n",
    "                gene_transcripts[ gene_id ] = []\n",
    "                gene_transcripts[ gene_id ].append( transcript_id )\n",
    "                                            \n",
    "        else:\n",
    "            pass\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "# parse for genes with just one hit have evalue 0.0\n",
    "gene_replacement = {}\n",
    "for next_transcript in transcript_genes.keys():\n",
    "\n",
    "    if len( transcript_genes[ next_transcript ] ) > 1:\n",
    "            \n",
    "        for next_gene in transcript_genes[ next_transcript ]:\n",
    "            gene_report_stats = transcript_report[ next_transcript ][ next_gene ]\n",
    "            output = ''\n",
    "            for next_stat in gene_report_stats:\n",
    "                output = output + next_stat + '\\t'\n",
    "            output = output[ :-1 ] + '\\n'\n",
    "            output_report_stats.write( output )\n",
    "\n",
    "    else:\n",
    "        gene_id = transcript_genes[ next_transcript ][ 0 ]\n",
    "        gene_sequence = genome_seq[ gene_id ]\n",
    "        transcript_sequence = transcriptome_seq[ next_transcript ]\n",
    "\n",
    "        gene_base = gene_id.split( '-' )[ 0 ]\n",
    "\n",
    "        if gene_base in gene_replacement.keys():\n",
    "            if len( transcriptome_seq[ next_transcript ] ) > len( gene_replacement[ gene_base ] ):\n",
    "                gene_replacement[ gene_base ] = next_transcript\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            gene_replacement[ gene_base ] = next_transcript\n",
    "\n",
    "# update genome gene model if there is a  pacbio sequence - assume pacbio sequence is better since read is real - gene model is hypothesized based on evidence of varying unknown specific quality\n",
    "# >ML49653a-RA cdna supercontig:MneLei_Aug2011:ML4965:35353:38358:-1 gene:ML49653a gene_biotype:protein_coding transcript_biotype:protein_coding_pacbioMnemiopsis-leidyi-NonamEVm004368t1\n",
    "\n",
    "genome_updated_letters = ''\n",
    "for next_gene in genome_seq.keys():\n",
    "    next_gene_base = next_gene.split( '-' )[ 0 ]\n",
    "    \n",
    "    if next_gene_base in gene_replacement.keys():\n",
    "        transcript_id = gene_replacement[ next_gene_base ]\n",
    "        next_gene_info = next_gene.split( ' ' )\n",
    "        transcript_info_1 = transcript_id.split( ' ' )\n",
    "        transcript_info_2 = transcript_info_1[ 0 ].split( '-' )\n",
    "        gene_id = next_gene_info[ 0 ]\n",
    "        original_seq = genome_seq[ gene_id ]\n",
    "        updated_gene_id = gene_id + '___pacbio' + transcript_info_2[ -1 ] #+ ' ' + ' '.join( next_gene_info[ 1: ] )\n",
    "        updated_seq = transcriptome_seq[ transcript_id ]\n",
    "\n",
    "        difference = str( len( updated_seq ) - len( original_seq ) )\n",
    "        output = 'PacBio ' + transcript_id + ' ' + str( len( updated_seq ) ) + '\\t Gene ' + gene_id + ' ' + str( len( original_seq ) ) + '\\t PacBio - Gene = ' + difference + ' nts CDS\\n'\n",
    "        output_stats.write( output )\n",
    "        output = str( len( updated_seq ) ) + '\\t' + str( len( original_seq ) ) + '\\t' + difference + '\\n'\n",
    "        output_difference.write( output)\n",
    "\n",
    "        output = '>' + updated_gene_id + '\\n' + updated_seq + '\\n'\n",
    "        output_fasta.write( output )\n",
    "        genome_updated_letters = genome_updated_letters + updated_seq\n",
    "\n",
    "        output = next_gene + '\\t' + transcript_id + '\\t' + updated_gene_id + '\\n'\n",
    "        output_map.write( output )\n",
    "        \n",
    "    else:\n",
    "        sequence = genome_seq[ next_gene ]\n",
    "        output = '>' + next_gene + '\\n' + sequence + '\\n'\n",
    "        output_fasta.write( output )\n",
    "        genome_updated_letters = genome_updated_letters + sequence\n",
    "\n",
    "output = 'Original gene models = ' + str( len( genome_original_letters ) ) + ' nucleotide CDS sequence'\n",
    "print( output)\n",
    "output_stats.write( output )\n",
    "output = 'Updated gene models = ' + str( len( genome_updated_letters ) ) + ' nucleotide CDS sequence'\n",
    "print( output)\n",
    "output_stats.write( output )\n",
    "output = 'Number of nucleotides added to CDS gene models = ' + str( len( genome_updated_letters ) - len( genome_original_letters ) )\n",
    "print( output )\n",
    "output_stats.write( output )\n",
    "\n",
    "input_report.close()\n",
    "input_transcriptome.close()\n",
    "input_genome.close()\n",
    "output_fasta.close()\n",
    "output_map.close()\n",
    "output_difference.close()\n",
    "output_stats.close()\n",
    "output_report_stats.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891faaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#008-python-update-gene-models-AA\n",
    "\n",
    "\n",
    "input_cds = open( 'output/7-Mnemiopsis_leidyi-genome-pacbio-updates.cds', 'r' )\n",
    "input_aa = open( 'output/3-Mnemiopsis-leidyi-pacbio-evigeneT1.aa', 'r' )\n",
    "input_genome_aa = open( 'output/Mnemiopsis_leidyi.MneLei_Aug2011.pep.all.fa', 'r' )\n",
    "output_aa = open( 'output/8-Mnemiopsis_leidyi-genome-pacbio-updates.aa', 'w' )\n",
    "\n",
    "aa_seq = {}\n",
    "# >Mnemiopsis-leidyi-NonamEVm000019t1\n",
    "for next_line in input_aa:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "\n",
    "        evigene_id = next_line[ 1:-1 ].split( '-' )[ -1 ]\n",
    "        aa_seq[ evigene_id ] = ''\n",
    "\n",
    "    else:\n",
    "\n",
    "        sequence = next_line[ :-1 ]\n",
    "        aa_seq[ evigene_id ] = aa_seq[ evigene_id ] + sequence\n",
    "\n",
    "cds_seq = {}\n",
    "# >ML00011a-RA___pacbioNonamEVm009938t1\n",
    "for next_line in input_cds:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "\n",
    "        identifier = next_line[ 1:-1 ]\n",
    "        cds_seq[ identifier ] = ''\n",
    "\n",
    "    else:\n",
    "\n",
    "        sequence = next_line[ :-1 ]\n",
    "        cds_seq[ identifier ] = cds_seq[ identifier ] + sequence\n",
    "\n",
    "genome_aa_seq = {}\n",
    "for next_line in input_genome_aa:\n",
    "\n",
    "    if next_line[ 0 ] == '>':\n",
    "\n",
    "        ensembl_id = next_line[ 1:-1 ].split( ' ' )[ 0 ]\n",
    "        genome_aa_seq[ ensembl_id ] = ''\n",
    "\n",
    "    else:\n",
    "\n",
    "        sequence = next_line[ :-1 ]\n",
    "        genome_aa_seq[ ensembl_id ] = genome_aa_seq[ ensembl_id ] + sequence\n",
    "        \n",
    "for next_id in cds_seq.keys():\n",
    "\n",
    "    info = next_id.split( '___' )\n",
    "\n",
    "    if len( info ) > 1:\n",
    "        evigene_id = info[ -1 ].split( 'bio' )[ -1 ]\n",
    "        sequence = aa_seq[ evigene_id ]\n",
    "        ensembl_id = info[ 0 ].split( '-' )[ 0 ] \n",
    "\n",
    "        output = '>' + ensembl_id  + '-PA___pacbio' + evigene_id + '\\n' + sequence + '\\n'\n",
    "        output_aa.write( output )\n",
    "\n",
    "    else:\n",
    "        info_id = next_id.split( '-' )\n",
    "        if len( info_id ) > 1:\n",
    "            ensembl_id = info_id[ 0 ] + '-PA'\n",
    "        else:\n",
    "            ensembl_id = info_id[ 0 ]\n",
    "\n",
    "        sequence = genome_aa_seq[ ensembl_id ]\n",
    "        output = '>' + ensembl_id + '\\n' + sequence + '\\n'\n",
    "        output_aa.write( output )\n",
    "            \n",
    "input_cds.close()\n",
    "input_aa.close()\n",
    "output_aa.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "879830dd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
