{
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh --quiet\n",
        "sha256sum Miniconda3-latest-Linux-x86_64.sh #should be 1314b90489f154602fd794accfc90446111514a5a72fe1f71ab83e07de9504a7\n",
        "bash Miniconda3-latest-Linux-x86_64.sh -b -p $HOME/miniconda3 > out.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o2NwsxRePYjO",
        "outputId": "86404108-8d89-4f46-f251-b4c91275810b"
      },
      "id": "o2NwsxRePYjO",
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "4ee9c3aa53329cd7a63b49877c0babb49b19b7e5af29807b793a76bdb1d362b4  Miniconda3-latest-Linux-x86_64.sh\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\r  0%|          | 0/39 [00:00<?, ?it/s]\rExtracting : wheel-0.37.1-pyhd3eb1b0_0.conda:   0%|          | 0/39 [00:00<?, ?it/s]\rExtracting : idna-3.3-pyhd3eb1b0_0.conda:   3%|▎         | 1/39 [00:00<00:05,  6.37it/s]\rExtracting : idna-3.3-pyhd3eb1b0_0.conda:   5%|▌         | 2/39 [00:00<00:02, 12.72it/s]\rExtracting : conda-content-trust-0.1.1-pyhd3eb1b0_0.conda:   5%|▌         | 2/39 [00:00<00:02, 12.72it/s]\rExtracting : cryptography-36.0.0-py39h9ce1e76_0.conda:   8%|▊         | 3/39 [00:00<00:02, 12.72it/s]    \rExtracting : ruamel_yaml-0.15.100-py39h27cfd23_0.conda:  10%|█         | 4/39 [00:00<00:02, 12.72it/s]\rExtracting : ruamel_yaml-0.15.100-py39h27cfd23_0.conda:  13%|█▎        | 5/39 [00:00<00:01, 19.03it/s]\rExtracting : brotlipy-0.7.0-py39h27cfd23_1003.conda:  13%|█▎        | 5/39 [00:00<00:01, 19.03it/s]   \rExtracting : libgcc-ng-9.3.0-h5101ec6_17.conda:  15%|█▌        | 6/39 [00:00<00:01, 19.03it/s]     \rExtracting : libgcc-ng-9.3.0-h5101ec6_17.conda:  18%|█▊        | 7/39 [00:00<00:01, 18.05it/s]\rExtracting : sqlite-3.37.0-hc218d9a_0.conda:  18%|█▊        | 7/39 [00:00<00:01, 18.05it/s]   \rExtracting : yaml-0.2.5-h7b6447c_0.conda:  21%|██        | 8/39 [00:00<00:01, 18.05it/s]   \rExtracting : six-1.16.0-pyhd3eb1b0_0.conda:  23%|██▎       | 9/39 [00:00<00:01, 18.05it/s]\rExtracting : readline-8.1.2-h7f8727e_1.conda:  26%|██▌       | 10/39 [00:00<00:01, 18.05it/s]\rExtracting : openssl-1.1.1m-h7f8727e_0.conda:  28%|██▊       | 11/39 [00:00<00:01, 18.05it/s]\rExtracting : openssl-1.1.1m-h7f8727e_0.conda:  31%|███       | 12/39 [00:00<00:01, 18.54it/s]\rExtracting : pysocks-1.7.1-py39h06a4308_0.conda:  31%|███       | 12/39 [00:00<00:01, 18.54it/s]\rExtracting : zlib-1.2.11-h7f8727e_4.conda:  33%|███▎      | 13/39 [00:00<00:01, 18.54it/s]      \rExtracting : urllib3-1.26.7-pyhd3eb1b0_0.conda:  36%|███▌      | 14/39 [00:00<00:01, 18.54it/s]\rExtracting : pip-21.2.4-py39h06a4308_0.conda:  38%|███▊      | 15/39 [00:00<00:01, 18.54it/s]  \rExtracting : requests-2.27.1-pyhd3eb1b0_0.conda:  41%|████      | 16/39 [00:00<00:01, 18.54it/s]\rExtracting : certifi-2021.10.8-py39h06a4308_2.conda:  44%|████▎     | 17/39 [00:00<00:01, 18.54it/s]\rExtracting : python-3.9.7-h12debd9_1.conda:  46%|████▌     | 18/39 [00:01<00:01, 18.54it/s]         \rExtracting : python-3.9.7-h12debd9_1.conda:  49%|████▊     | 19/39 [00:01<00:02,  9.17it/s]\rExtracting : xz-5.2.5-h7b6447c_0.conda:  49%|████▊     | 19/39 [00:01<00:02,  9.17it/s]    \rExtracting : ld_impl_linux-64-2.35.1-h7274673_9.conda:  51%|█████▏    | 20/39 [00:01<00:02,  9.17it/s]\rExtracting : conda-package-handling-1.7.3-py39h27cfd23_1.conda:  54%|█████▍    | 21/39 [00:01<00:01,  9.17it/s]\rExtracting : ca-certificates-2021.10.26-h06a4308_2.conda:  56%|█████▋    | 22/39 [00:01<00:01,  9.17it/s]      \rExtracting : pycosat-0.6.3-py39h27cfd23_0.conda:  59%|█████▉    | 23/39 [00:01<00:01,  9.17it/s]         \rExtracting : charset-normalizer-2.0.4-pyhd3eb1b0_0.conda:  62%|██████▏   | 24/39 [00:01<00:01,  9.17it/s]\rExtracting : libgomp-9.3.0-h5101ec6_17.conda:  64%|██████▍   | 25/39 [00:01<00:01,  9.17it/s]            \rExtracting : libffi-3.3-he6710b0_2.conda:  67%|██████▋   | 26/39 [00:01<00:01,  9.17it/s]    \rExtracting : conda-4.11.0-py39h06a4308_0.conda:  69%|██████▉   | 27/39 [00:02<00:01,  9.17it/s]\rExtracting : conda-4.11.0-py39h06a4308_0.conda:  72%|███████▏  | 28/39 [00:02<00:00, 14.22it/s]\rExtracting : _libgcc_mutex-0.1-main.conda:  72%|███████▏  | 28/39 [00:02<00:00, 14.22it/s]     \rExtracting : cffi-1.15.0-py39hd667e15_1.conda:  74%|███████▍  | 29/39 [00:02<00:00, 14.22it/s]\rExtracting : tk-8.6.11-h1ccaba5_0.conda:  77%|███████▋  | 30/39 [00:02<00:00, 14.22it/s]      \rExtracting : libstdcxx-ng-9.3.0-hd4cf53a_17.conda:  79%|███████▉  | 31/39 [00:02<00:00, 14.22it/s]\rExtracting : setuptools-58.0.4-py39h06a4308_0.conda:  82%|████████▏ | 32/39 [00:02<00:00, 14.22it/s]\rExtracting : setuptools-58.0.4-py39h06a4308_0.conda:  85%|████████▍ | 33/39 [00:02<00:00, 16.35it/s]\rExtracting : ncurses-6.3-h7f8727e_2.conda:  85%|████████▍ | 33/39 [00:02<00:00, 16.35it/s]          \rExtracting : tqdm-4.62.3-pyhd3eb1b0_1.conda:  87%|████████▋ | 34/39 [00:02<00:00, 16.35it/s]\rExtracting : pyopenssl-21.0.0-pyhd3eb1b0_1.conda:  90%|████████▉ | 35/39 [00:02<00:00, 16.35it/s]\rExtracting : pyopenssl-21.0.0-pyhd3eb1b0_1.conda:  92%|█████████▏| 36/39 [00:02<00:00, 16.53it/s]\rExtracting : tzdata-2021e-hda174b7_0.conda:  92%|█████████▏| 36/39 [00:02<00:00, 16.53it/s]      \rExtracting : pycparser-2.21-pyhd3eb1b0_0.conda:  95%|█████████▍| 37/39 [00:02<00:00, 16.53it/s]\rExtracting : _openmp_mutex-4.5-1_gnu.tar.bz2:  97%|█████████▋| 38/39 [00:02<00:00, 16.53it/s]  \r                                                                                             \r"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "source $HOME/miniconda3/bin/activate\n",
        "conda install -c bioconda blast"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0TM-gh80Pcik",
        "outputId": "8e413d40-7da8-4cd6-a60b-bc7181a50d1c"
      },
      "id": "0TM-gh80Pcik",
      "execution_count": 49,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "# All requested packages already installed.\n",
            "\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/miniconda3\n",
            "\n",
            "  added / updated specs:\n",
            "    - blast\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    pyopenssl-22.0.0           |     pyhd3eb1b0_0          50 KB\n",
            "    six-1.16.0                 |     pyhd3eb1b0_1          18 KB\n",
            "    sqlite-3.37.2              |       hc218d9a_0        1008 KB\n",
            "    urllib3-1.26.8             |     pyhd3eb1b0_0         106 KB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         1.2 MB\n",
            "\n",
            "The following packages will be REMOVED:\n",
            "\n",
            "  conda-content-trust-0.1.1-pyhd3eb1b0_0\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  pyopenssl                             21.0.0-pyhd3eb1b0_1 --> 22.0.0-pyhd3eb1b0_0\n",
            "  six                                   1.16.0-pyhd3eb1b0_0 --> 1.16.0-pyhd3eb1b0_1\n",
            "  sqlite                                  3.37.0-hc218d9a_0 --> 3.37.2-hc218d9a_0\n",
            "  urllib3                               1.26.7-pyhd3eb1b0_0 --> 1.26.8-pyhd3eb1b0_0\n",
            "\n",
            "\n",
            "Proceed ([y]/n)? \n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\rpyopenssl-22.0.0     | 50 KB     |            |   0% \rpyopenssl-22.0.0     | 50 KB     | ###2       |  32% \rpyopenssl-22.0.0     | 50 KB     | ########## | 100% \n",
            "\rsqlite-3.37.2        | 1008 KB   |            |   0% \rsqlite-3.37.2        | 1008 KB   | ########## | 100% \rsqlite-3.37.2        | 1008 KB   | ########## | 100% \n",
            "\rsix-1.16.0           | 18 KB     |            |   0% \rsix-1.16.0           | 18 KB     | ########## | 100% \n",
            "\rurllib3-1.26.8       | 106 KB    |            |   0% \rurllib3-1.26.8       | 106 KB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "source $HOME/miniconda3/bin/activate\n",
        "conda install -c bioconda mafft=7.487"
      ],
      "metadata": {
        "id": "keWN4xErWNrj",
        "outputId": "8a83025d-d28e-4cd1-ecaa-e1a3baa21497",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "keWN4xErWNrj",
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/miniconda3\n",
            "\n",
            "  added / updated specs:\n",
            "    - mafft=7.487\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    mafft-7.487                |       h779adbc_0         3.0 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         3.0 MB\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  mafft                                    7.310-h1b792b2_4 --> 7.487-h779adbc_0\n",
            "\n",
            "\n",
            "Proceed ([y]/n)? \n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\rmafft-7.487          | 3.0 MB    |            |   0% \rmafft-7.487          | 3.0 MB    |            |   1% \rmafft-7.487          | 3.0 MB    | ########## | 100% \rmafft-7.487          | 3.0 MB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "source $HOME/miniconda3/bin/activate\n",
        "conda install -c bioconda clipkit"
      ],
      "metadata": {
        "id": "mKhNNSU0WiDp",
        "outputId": "caf2ed18-f48b-44b5-e6a1-abf4adfcc9f1",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "mKhNNSU0WiDp",
      "execution_count": 42,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/miniconda3\n",
            "\n",
            "  added / updated specs:\n",
            "    - clipkit\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    biopython-1.78             |   py39h7f8727e_0         2.2 MB\n",
            "    blas-1.0                   |              mkl           6 KB\n",
            "    clipkit-1.3.0              |     pyhdfd78af_0          18 KB  bioconda\n",
            "    intel-openmp-2021.4.0      |    h06a4308_3561         4.2 MB\n",
            "    mkl-2021.4.0               |     h06a4308_640       142.6 MB\n",
            "    mkl-service-2.4.0          |   py39h7f8727e_0          59 KB\n",
            "    mkl_fft-1.3.1              |   py39hd3c417c_0         182 KB\n",
            "    mkl_random-1.2.2           |   py39h51133e4_0         309 KB\n",
            "    numpy-1.21.2               |   py39h20f2e39_0          23 KB\n",
            "    numpy-base-1.21.2          |   py39h79a1101_0         4.9 MB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       154.4 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  biopython          pkgs/main/linux-64::biopython-1.78-py39h7f8727e_0\n",
            "  blas               pkgs/main/linux-64::blas-1.0-mkl\n",
            "  clipkit            bioconda/noarch::clipkit-1.3.0-pyhdfd78af_0\n",
            "  intel-openmp       pkgs/main/linux-64::intel-openmp-2021.4.0-h06a4308_3561\n",
            "  mkl                pkgs/main/linux-64::mkl-2021.4.0-h06a4308_640\n",
            "  mkl-service        pkgs/main/linux-64::mkl-service-2.4.0-py39h7f8727e_0\n",
            "  mkl_fft            pkgs/main/linux-64::mkl_fft-1.3.1-py39hd3c417c_0\n",
            "  mkl_random         pkgs/main/linux-64::mkl_random-1.2.2-py39h51133e4_0\n",
            "  numpy              pkgs/main/linux-64::numpy-1.21.2-py39h20f2e39_0\n",
            "  numpy-base         pkgs/main/linux-64::numpy-base-1.21.2-py39h79a1101_0\n",
            "\n",
            "\n",
            "Proceed ([y]/n)? \n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\rmkl-service-2.4.0    | 59 KB     |            |   0% \rmkl-service-2.4.0    | 59 KB     | ########1  |  81% \rmkl-service-2.4.0    | 59 KB     | ########## | 100% \n",
            "\rmkl_fft-1.3.1        | 182 KB    |            |   0% \rmkl_fft-1.3.1        | 182 KB    | ########## | 100% \rmkl_fft-1.3.1        | 182 KB    | ########## | 100% \n",
            "\rmkl-2021.4.0         | 142.6 MB  |            |   0% \rmkl-2021.4.0         | 142.6 MB  | 1          |   1% \rmkl-2021.4.0         | 142.6 MB  | 8          |   9% \rmkl-2021.4.0         | 142.6 MB  | #3         |  13% \rmkl-2021.4.0         | 142.6 MB  | #8         |  19% \rmkl-2021.4.0         | 142.6 MB  | ##5        |  25% \rmkl-2021.4.0         | 142.6 MB  | ###        |  30% \rmkl-2021.4.0         | 142.6 MB  | ###5       |  35% \rmkl-2021.4.0         | 142.6 MB  | ####       |  41% \rmkl-2021.4.0         | 142.6 MB  | ####5      |  46% \rmkl-2021.4.0         | 142.6 MB  | #####      |  51% \rmkl-2021.4.0         | 142.6 MB  | #####5     |  55% \rmkl-2021.4.0         | 142.6 MB  | ######2    |  63% \rmkl-2021.4.0         | 142.6 MB  | ######8    |  69% \rmkl-2021.4.0         | 142.6 MB  | #######4   |  75% \rmkl-2021.4.0         | 142.6 MB  | ########   |  80% \rmkl-2021.4.0         | 142.6 MB  | ########5  |  85% \rmkl-2021.4.0         | 142.6 MB  | #########1 |  92% \rmkl-2021.4.0         | 142.6 MB  | #########6 |  97% \rmkl-2021.4.0         | 142.6 MB  | ########## | 100% \n",
            "\rnumpy-base-1.21.2    | 4.9 MB    |            |   0% \rnumpy-base-1.21.2    | 4.9 MB    | ########## | 100% \rnumpy-base-1.21.2    | 4.9 MB    | ########## | 100% \n",
            "\rintel-openmp-2021.4. | 4.2 MB    |            |   0% \rintel-openmp-2021.4. | 4.2 MB    | ########## | 100% \rintel-openmp-2021.4. | 4.2 MB    | ########## | 100% \n",
            "\rnumpy-1.21.2         | 23 KB     |            |   0% \rnumpy-1.21.2         | 23 KB     | ########## | 100% \n",
            "\rclipkit-1.3.0        | 18 KB     |            |   0% \rclipkit-1.3.0        | 18 KB     | ########9  |  90% \rclipkit-1.3.0        | 18 KB     | ########## | 100% \n",
            "\rbiopython-1.78       | 2.2 MB    |            |   0% \rbiopython-1.78       | 2.2 MB    | ########## | 100% \rbiopython-1.78       | 2.2 MB    | ########## | 100% \n",
            "\rblas-1.0             | 6 KB      |            |   0% \rblas-1.0             | 6 KB      | ########## | 100% \n",
            "\rmkl_random-1.2.2     | 309 KB    |            |   0% \rmkl_random-1.2.2     | 309 KB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "source $HOME/miniconda3/bin/activate\n",
        "conda install -c bioconda iqtree"
      ],
      "metadata": {
        "id": "wm5j5JJWZn0-",
        "outputId": "b09ce500-128d-4676-9bbf-1239a83fa3e6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "id": "wm5j5JJWZn0-",
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/miniconda3\n",
            "\n",
            "  added / updated specs:\n",
            "    - iqtree\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    iqtree-2.0.3               |       h176a8bc_1         3.0 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         3.0 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  iqtree             bioconda/linux-64::iqtree-2.0.3-h176a8bc_1\n",
            "\n",
            "\n",
            "Proceed ([y]/n)? \n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\riqtree-2.0.3         | 3.0 MB    |            |   0% \riqtree-2.0.3         | 3.0 MB    |            |   1% \riqtree-2.0.3         | 3.0 MB    | ######     |  61% \riqtree-2.0.3         | 3.0 MB    | ########## | 100% \riqtree-2.0.3         | 3.0 MB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "### No Need to Modify ###\n",
        "pip install gdown --upgrade --no-cache-dir --quiet &&\n",
        "gdown --id 11S8ROPmgjhIakLOcBhz-SJ2q0-d3wTrV &&\n",
        "tar -xf 6-homologs-genes.tar.gz\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4SqKTRRHNR9H",
        "outputId": "7c9d4c2e-1255-44f8-f03f-9d9b8331c3d0"
      },
      "id": "4SqKTRRHNR9H",
      "execution_count": 61,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=11S8ROPmgjhIakLOcBhz-SJ2q0-d3wTrV\n",
            "To: /content/6-homologs-genes/6-homologs-genes.tar.gz\n",
            "\r  0%|          | 0.00/181M [00:00<?, ?B/s]\r  0%|          | 524k/181M [00:00<00:46, 3.90MB/s]\r  5%|▍         | 8.91M/181M [00:00<00:03, 44.6MB/s]\r  9%|▉         | 16.3M/181M [00:00<00:02, 55.7MB/s]\r 12%|█▏        | 22.5M/181M [00:00<00:02, 54.5MB/s]\r 16%|█▌        | 28.3M/181M [00:00<00:03, 45.8MB/s]\r 22%|██▏       | 39.8M/181M [00:00<00:02, 63.1MB/s]\r 27%|██▋       | 48.2M/181M [00:00<00:01, 68.9MB/s]\r 31%|███       | 55.6M/181M [00:00<00:01, 69.0MB/s]\r 35%|███▍      | 62.9M/181M [00:01<00:01, 67.2MB/s]\r 41%|████      | 74.4M/181M [00:01<00:01, 80.0MB/s]\r 46%|████▌     | 82.8M/181M [00:01<00:01, 74.9MB/s]\r 50%|█████     | 90.7M/181M [00:01<00:01, 74.4MB/s]\r 55%|█████▍    | 99.1M/181M [00:01<00:01, 76.3MB/s]\r 59%|█████▉    | 107M/181M [00:01<00:00, 75.9MB/s] \r 64%|██████▍   | 116M/181M [00:01<00:00, 80.5MB/s]\r 69%|██████▉   | 125M/181M [00:01<00:00, 79.3MB/s]\r 78%|███████▊  | 141M/181M [00:01<00:00, 101MB/s] \r 84%|████████▎ | 152M/181M [00:02<00:00, 85.3MB/s]\r 91%|█████████ | 164M/181M [00:02<00:00, 95.2MB/s]\r 96%|█████████▋| 175M/181M [00:02<00:00, 94.5MB/s]\r100%|██████████| 181M/181M [00:02<00:00, 76.4MB/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "id": "5cb87672",
      "metadata": {
        "id": "5cb87672"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "os.chdir('6-homologs-genes')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "f1cabedc",
      "metadata": {
        "id": "f1cabedc"
      },
      "outputs": [],
      "source": [
        "#000-group-basal-metazoans\n",
        "\n",
        "# Nematostella vectensis\n",
        "# Morbakka virulenta\n",
        "# Hydra vulgaris\n",
        "# Thelohanellus kitauei\n",
        "# Rhopilema esculentum\n",
        "# Hormiphora californensis\n",
        "# Mnemiopsis leidyi\n",
        "# Trichoplax adhaerens\n",
        "# Ephydatia muelleri\n",
        "# Amphimedon queenslandica\n",
        "# Monosiga brevicollis\n",
        "# Capsaspora owczarzaki\n",
        "# Sphaeroforma arctica\n",
        "# Homo sapiens\n",
        "# Drosophila melanogaster\n",
        "# Caenorhabditis elegans\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "id": "a27773be",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a27773be",
        "outputId": "1095ab9a-0acf-4ed0-be8c-328f0f807b46"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#001-ls-blastdbs\n",
        "\n",
        "mkdir output || echo \"\"\n",
        "echo\n",
        "ls -1 4-projectdb-projectdb/projectdb-fastas-provided/*aa > output/1-list-projectdb-blastdbs\n",
        "ls -1 source-rgs/rgs5-piezo-full_length-03dec2021.fasta > output/1-list-rgs\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "id": "5626c364",
      "metadata": {
        "id": "5626c364"
      },
      "outputs": [],
      "source": [
        "#002-python-command-blastp\n",
        "\n",
        "# input_list\n",
        "# localdb/projectdb-Metazoa-Nematoda-Chromadorea-Rhabditida-Onchocercidae-Brugia-malayi-6279-Bmal-4.0.fasta\n",
        "\n",
        "input_list = open( 'output/1-list-projectdb-blastdbs', 'r' )\n",
        "input_rgs = open( 'output/1-list-rgs', 'r' )\n",
        "output_command = open( '003-blastp_X_projectDB', 'w' )\n",
        "\n",
        "for next_rgs in input_rgs:\n",
        "\n",
        "    rgs_fasta = next_rgs[ :-1 ] # ends up assigning to pore region in this case\n",
        "\n",
        "for next_line in input_list:\n",
        "\n",
        "    db_path = next_line[ :-1 ]\n",
        "    info = next_line[ :-1 ].split( '-' )\n",
        "    genus = info[ 10 ]\n",
        "    species = info[ 11 ]\n",
        "    gspp = genus + '-' + species\n",
        "    \n",
        "    output = 'blastp -db ' + db_path  + ' -outfmt 6 -out output/3-blast-report-rgs_X_projectdb-' + gspp + ' -max_hsps 1 -query ' + rgs_fasta + ' -evalue 1e-3\\n'\n",
        "    output_command.write( output )\n",
        "    \n",
        "output = 'echo\\n'\n",
        "output_command.write( output )\n",
        "\n",
        "input_list.close()\n",
        "input_rgs.close()\n",
        "output_command.close()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "id": "0f9465ce",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0f9465ce",
        "outputId": "79aa43aa-2a03-480c-86e1-33427071baa2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#003-blastp_X_projectDB\n",
        "source $HOME/miniconda3/bin/activate\n",
        "bash 003-blastp_X_projectDB\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "#004-ls-reports-and-fastas\n",
        "\n",
        "ls output/3-blast-report-rgs_X_projectdb-* > output/4-list-reports\n",
        "ls 4-projectdb-projectdb/projectdb-fastas-provided/*aa > output/4-list-fastas\n"
      ],
      "metadata": {
        "id": "57ThZxoXSLLj"
      },
      "id": "57ThZxoXSLLj",
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "id": "ea606dd9",
      "metadata": {
        "id": "ea606dd9"
      },
      "outputs": [],
      "source": [
        "#005-python-gene-set-fasta\n",
        "\n",
        "# rgs73-worm-TRPA_Cele_TRPA-TRPA2_1_481-uniprotQ21517-extraction_51_366   Metazoa-Chordata-Coelacanthimorpha-Coelacanthiformes-Coelacanthidae-Latimeria-chalumnae-pdb0000292542   26.036  169     98      4       140     299     846     996     3.39e-05        44.4\n",
        "\n",
        "input_fastas = open( 'output/4-list-fastas', 'r' )\n",
        "input_reports = open( 'output/4-list-reports', 'r' )\n",
        "output_geneset_gspp = open( 'output/5-blastp-hits.fasta', 'w')\n",
        "output_geneset_gspp_sub = open( 'output/5-blastp-pore-region-hits.fasta', 'w')\n",
        "\n",
        "gene_seq = {}\n",
        "gene_coordinates = {}\n",
        "for next_fasta in input_fastas:\n",
        "    \n",
        "    fasta = next_fasta[ :-1 ]\n",
        "    input_fasta = open( fasta, 'r' )\n",
        "    \n",
        "    for next_line in input_fasta:\n",
        "\n",
        "        if next_line[ 0 ] == '>':\n",
        "            \n",
        "            header = next_line[ 1:-1 ]\n",
        "            gene_seq[ header ] = ''\n",
        "\n",
        "        else:\n",
        "\n",
        "            sequence = next_line[ :-1 ]\n",
        "            gene_seq[ header ] = gene_seq[ header ] + sequence\n",
        "\n",
        "    input_fasta.close()\n",
        "\n",
        "for next_report in input_reports:\n",
        "\n",
        "    all_hits = []\n",
        "    report = next_report[ :-1 ]\n",
        "    input_report = open( report, 'r' )\n",
        " \n",
        "    for next_line in input_report:\n",
        "\n",
        "        report_info = next_line.split( '\\t' )\n",
        "        gene_id = report_info[ 1 ]\n",
        "        all_hits.append( gene_id )\n",
        "        coordinate_1 = int( report_info[ 8 ] ) - 1\n",
        "        coordinate_2 = int( report_info[ 9 ] ) - 1\n",
        "        gene_coordinates[ gene_id ] = ( coordinate_1, coordinate_2 )\n",
        "        \n",
        "    unique_hits = list( set( all_hits ) )\n",
        "\n",
        "    for next_hit in unique_hits:\n",
        "        \n",
        "        gene_id = next_hit\n",
        "        sequence = gene_seq[ gene_id ]\n",
        "        coordinate_1 = gene_coordinates[ gene_id ][ 0 ]\n",
        "        coordinate_2 = gene_coordinates[ gene_id ][ 1 ]\n",
        "        subsequence = sequence[ coordinate_1 : coordinate_2 ] \n",
        "\n",
        "        output = '>' + gene_id + '\\n' + sequence + '\\n'\n",
        "        output_geneset_gspp.write( output )\n",
        "        \n",
        "        output_sub = '>' + gene_id + '\\n' + subsequence + '\\n'\n",
        "        output_geneset_gspp_sub.write( output_sub )\n",
        "\n",
        "    input_report.close()\n",
        "\n",
        "input_fastas.close()\n",
        "input_reports.close()\n",
        "output_geneset_gspp.close()\n",
        "output_geneset_gspp_sub.close()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "id": "8216c566",
      "metadata": {
        "id": "8216c566"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#006-blastp-rgs_X_rgs-genomes\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "blastp -db 4-projectdb-projectdb/projectdb-fastas-provided/projectdb-Metazoa-Chordata-Mammalia-Primates-Hominidae-Homo-sapiens*aa -query source-rgs/rgs5-piezo-full_length-03dec2021.fasta -out output/6-blast-report-rgs5-piezo-family_X_human-genome -outfmt 6  -max_hsps 1\n",
        "\n",
        "blastp -db 4-projectdb-projectdb/projectdb-fastas-provided/projectdb-Metazoa-Arthropoda-Insecta-Diptera-Drosophilidae-Drosophila-melanogaster*aa -query source-rgs/rgs5-piezo-full_length-03dec2021.fasta -out output/6-blast-report-rgs5-piezo-family_X_fly-genome -outfmt 6  -max_hsps 1\n",
        "\n",
        "blastp -db 4-projectdb-projectdb/projectdb-fastas-provided/projectdb-Metazoa-Nematoda-Chromadorea-Rhabditida-Rhabditidae-Caenorhabditis-elegans*aa -query source-rgs/rgs5-piezo-full_length-03dec2021.fasta -out output/6-blast-report-rgs5-piezo-family_X_worm-genome -outfmt 6  -max_hsps 1\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "id": "c9cbad90",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c9cbad90",
        "outputId": "76e550fb-166f-451b-d7e8-ad66aeb70822"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ls: cannot access 'output/6*genome*': No such file or directory\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#007-ls-rgs-reports-fastas\n",
        "\n",
        "ls output/6*genome* > output/7-list-reports\n",
        "\n",
        "ls 4-projectdb-projectdb/projectdb-fastas-provided/*Homo-sapiens*aa > output/7-list-rgs-projectdb-fastas\n",
        "ls 4-projectdb-projectdb/projectdb-fastas-provided/*Drosophila-melanogaster*aa >> output/7-list-rgs-projectdb-fastas\n",
        "ls 4-projectdb-projectdb/projectdb-fastas-provided/*Caenorhabditis-elegans*aa >> output/7-list-rgs-projectdb-fastas\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "id": "9dcc57bc",
      "metadata": {
        "id": "9dcc57bc"
      },
      "outputs": [],
      "source": [
        "#008-python-update-reference-genomes\n",
        "\n",
        "\n",
        "##### USER INPUT\n",
        "input_reports = open( 'output/7-list-reports', 'r' )\n",
        "input_fastas = open( 'output/7-list-rgs-projectdb-fastas', 'r' )\n",
        "input_rgsfasta = open( 'source-rgs/rgs5-piezo-full_length-03dec2021.fasta', 'r' )\n",
        "output_map = open( 'output/8-map-source-to-reference-identifiers', 'a' )\n",
        "\n",
        "model_species = [ 'human','mouse','fly','worm','anemone' ]\n",
        "\n",
        "###### BEING SCRIPT \n",
        "# read in reference gene header identifier and sequence into dictionary\n",
        "rgs_seq = {}\n",
        "for next_line in input_rgsfasta:\n",
        "\n",
        "    if next_line[ 0 ] == '>':\n",
        "\n",
        "        identifier = next_line[ 1:-1 ].split( ' ' )[ 0 ]\n",
        "        rgs_seq[ identifier ] = ''\n",
        "\n",
        "    else:\n",
        "\n",
        "        rgs_seq[ identifier ] = rgs_seq[ identifier ] + next_line[ :-1 ]\n",
        "\n",
        "# read rgs query and rgs genome top hit into dictionary\n",
        "gengene_refgene = {}\n",
        "rgs_genes = []\n",
        "gengenes = []\n",
        "\n",
        "for next_report in input_reports:\n",
        "    \n",
        "    input_report = open( next_report[ :-1 ], 'r' )\n",
        "    \n",
        "    model_name = \"\"\n",
        "    for model in model_species:\n",
        "        if model in next_report:\n",
        "            model_name = model\n",
        "            break\n",
        "\n",
        "    for next_hit in input_report:\n",
        "\n",
        "        info = next_hit.split( '\\t' )\n",
        "        refgene = info[ 0 ]\n",
        "        gengene = info[ 1 ]\n",
        "        name = refgene.split( '-' )[ 1 ]\n",
        "        \n",
        "        if name == model_name:\n",
        "\n",
        "            if (refgene in rgs_genes) or (gengene in gengenes):\n",
        "\n",
        "                pass\n",
        "\n",
        "            else:\n",
        "\n",
        "                gengene_refgene[ gengene ] = refgene\n",
        "                rgs_genes.append( refgene )\n",
        "                gengenes.append( gengene )\n",
        "\n",
        "        else:\n",
        "\n",
        "            pass\n",
        "            \n",
        "    input_report.close()\n",
        "\n",
        "\n",
        "# read in RGS genome and replace rgs genes (top hit in blast of rgs _X_ rgs genome) with rgs header and sequence\n",
        "header_seq = {}\n",
        "for next_fasta in input_fastas:\n",
        "\n",
        "    input_fasta = open( next_fasta[ :-1 ], 'r' )\n",
        "    output_name = 'output/8-' + next_fasta[ :-1 ].split( '/' )[ -1 ] + '-rgs'\n",
        "    output_fasta = open( output_name, 'w' )\n",
        "\n",
        "    for next_line in input_fasta:\n",
        "\n",
        "        if next_line[ 0 ] == '>':\n",
        "\n",
        "            count = 0\n",
        "            \n",
        "            header_info = next_line[ 1:-1 ]\n",
        "\n",
        "            if header_info in gengene_refgene.keys():\n",
        "                \n",
        "                count = 1\n",
        "                gengene = header_info\n",
        "                refgene = gengene_refgene[ gengene ]\n",
        "                \n",
        "                header = '>' + refgene + '\\n'\n",
        "                output_fasta.write( header )\n",
        "\n",
        "                \n",
        "                output = gengene + '\\t' + refgene + '\\n'\n",
        "                output_map.write( output )                \n",
        "                \n",
        "            else:\n",
        "\n",
        "                header = next_line\n",
        "                output_fasta.write( header )\n",
        "                \n",
        "        else:\n",
        "\n",
        "            if count == 0:\n",
        "\n",
        "                sequence = next_line\n",
        "                output_fasta.write( sequence )\n",
        "\n",
        "            else:\n",
        "\n",
        "                sequence = rgs_seq[ refgene ] + '\\n'\n",
        "                output_fasta.write( sequence )\n",
        "\n",
        "    input_fasta.close()\n",
        "    output_fasta.close()\n",
        "\n",
        "input_reports.close()\n",
        "input_fastas.close()\n",
        "input_rgsfasta.close()\n",
        "output_map.close()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "id": "3f096158",
      "metadata": {
        "id": "3f096158"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#009-ls-reference-genome-fastas\n",
        "\n",
        "cat output/8*rgs > output/9-rgs-all-genomes-combined.fasta-rgs\n",
        "ls -1 output/9*rgs > output/9-list-RGS-header-reference-genome-fastas\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "id": "f74206a6",
      "metadata": {
        "id": "f74206a6"
      },
      "outputs": [],
      "source": [
        "#010-python-command-makeblastdb\n",
        "\n",
        "input_list = open( 'output/9-list-RGS-header-reference-genome-fastas', 'r' )\n",
        "output_makedb = open( '011-blastp-makedb', 'w' )\n",
        "\n",
        "for next_fasta in input_list:\n",
        "\n",
        "    fasta_path = next_fasta[ :-1 ]\n",
        "    fasta = fasta_path.split( '/' )[ -1 ]\n",
        "    \n",
        "    db_name = fasta.split( '.' )[ 0 ] + '-AA'\n",
        "    \n",
        "    command = 'makeblastdb -in '  + fasta_path + ' -parse_seqids -dbtype prot'\n",
        "    output = command + '\\n'\n",
        "    output_makedb.write( output )\n",
        "\n",
        "input_list.close()\n",
        "output_makedb.close()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 40,
      "id": "fc2a56a9",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fc2a56a9",
        "outputId": "16c51468-287e-4ec0-c907-4d8511912649"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 02/18/2022 05:41:28\n",
            "New DB name:   /content/6-homologs-genes/output/9-rgs-all-genomes-combined.fasta-rgs\n",
            "New DB title:  output/9-rgs-all-genomes-combined.fasta-rgs\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 53595 sequences in 2.84786 seconds.\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#011-blastp-makedb\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "bash 011-blastp-makedb"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "id": "2780794f",
      "metadata": {
        "id": "2780794f"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#012-ls-blastp-rgs-genomes\n",
        "\n",
        "ls -1 output/9-rgs-all-genomes-combined.fasta-rgs > output/12-list-blastp-annotated-rgs-genomes\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#013-python-command-blastp\n",
        "\n",
        "# input_list\n",
        "# 11-projectdb-Homo-sapiens-cgs-AA.dmnd\n",
        "\n",
        "input_list = open( 'output/12-list-blastp-annotated-rgs-genomes', 'r' )\n",
        "output_command = open( '014-blastp-hits_X_RGS-genomes', 'w' )\n",
        "\n",
        "output = '#! /bin/bash\\n'\n",
        "output_command.write( output )\n",
        "\n",
        "for next_line in input_list:\n",
        "\n",
        "    db_path = next_line[ :-1 ]\n",
        "    info = db_path.split( '-' )\n",
        "    genome = info[ -3 ] + '-' + info[ -2 ]\n",
        "\n",
        "    output =  'blastp -db ' + db_path  + ' -outfmt 6 -out output/14-blastp-report-blastp_hits_X_RGS-genome-' + genome + '  -max_target_seqs 1 -max_hsps 1 -query output/5-blastp-pore-region-hits.fasta -matrix BLOSUM45 -evalue 1e-3 -num_threads 60 &\\n'\n",
        " \n",
        "    output_command.write( output )\n",
        "\n",
        "input_list.close()\n",
        "output_command.close()\n"
      ],
      "metadata": {
        "id": "5QgaKQZeSpAr"
      },
      "id": "5QgaKQZeSpAr",
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "#014-blastp-hits_X_RGS-genomes\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "bash 014-blastp-hits_X_RGS-genomes"
      ],
      "metadata": {
        "id": "rvHCfCZZSqZl"
      },
      "id": "rvHCfCZZSqZl",
      "execution_count": 41,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 43,
      "id": "ce056c15",
      "metadata": {
        "id": "ce056c15"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#015-cat-all-blastp-reports\n",
        "\n",
        "cat output/14-* > output/15-all-blastp-all-reports\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "id": "820152e6",
      "metadata": {
        "id": "820152e6"
      },
      "outputs": [],
      "source": [
        "#016-python-RBF-CGS-each-RGS-genome-fasta\n",
        "\n",
        "\n",
        "##### USER INPUT\n",
        "\n",
        "input_fastas = open( 'output/4-list-fastas', 'r' )\n",
        "input_hits = open( 'output/15-all-blastp-all-reports', 'r' )\n",
        "input_rgs_ids = open( 'output/8-map-source-to-reference-identifiers', 'r' )\n",
        "\n",
        "output_fasta = open( 'output/16-CGS-final-sequences-by-blastp-RBF.fasta', 'w' )\n",
        "output_filtered = open( 'output/16-dropped-queries-no-rgs-top-hit-in-rgs-genome', 'w' )\n",
        "\n",
        "model_species = [ 'human','mouse','fly','worm','anemone' ]\n",
        "\n",
        "##### BEGIN SCRIPT\n",
        "\n",
        "rgs_ids = []\n",
        "\n",
        "for next_line in input_rgs_ids:\n",
        "\n",
        "        info = next_line[ :-1 ].split( '\\t' )\n",
        "        projectdb_id = info[ 0 ]\n",
        "        rgs_id = info[ 1 ]\n",
        "        rgs_ids.append( rgs_id )\n",
        "        rgs_ids.append( projectdb_id )\n",
        "\n",
        "keepers = []\n",
        "queries = []\n",
        "\n",
        "for next_hit in input_hits:\n",
        "\n",
        "        info = next_hit.split( '\\t' )\n",
        "        query = info[ 0 ]\n",
        "        queries.append( query )\n",
        "        \n",
        "        hit = info[ 1 ]        \n",
        "        hit_info = hit.split( '-' )\n",
        "        name = hit_info[ 1 ]\n",
        "\n",
        "        if name in model_species:\n",
        "                \n",
        "                # drop RGS genes\n",
        "                if query in rgs_ids:\n",
        "\n",
        "                        pass\n",
        "\n",
        "                else:\n",
        "                        keepers.append( query )\n",
        "\n",
        "        else:\n",
        "                output = query + '\\t' + hit + '\\n'\n",
        "                output_filtered.write( output )\n",
        "\n",
        "# produce post-filtered keepers diamond fasta NO RGS\n",
        "count = 0\n",
        "\n",
        "for next_fasta in input_fastas:\n",
        "\n",
        "        input_fasta = open( next_fasta[ :-1 ], 'r' )\n",
        "\n",
        "        for next_line in input_fasta:\n",
        "                \n",
        "                if next_line[ 0 ] == '>':\n",
        "                        count = 0\n",
        "                        identifier = next_line[ 1:-1 ]\n",
        "\n",
        "                        if identifier in keepers:\n",
        "                                count = 1\n",
        "                                header = '>' + identifier + '\\n'\n",
        "                                output_fasta.write( header )\n",
        "                                \n",
        "                        else:\n",
        "                                pass\n",
        "\n",
        "                else:\n",
        "\n",
        "                        if count == 0:\n",
        "                                pass\n",
        "\n",
        "                        else:\n",
        "\n",
        "                                sequence = next_line\n",
        "                                output_fasta.write( sequence )\n",
        "                                \n",
        "        input_fasta.close()\n",
        "            \n",
        "input_hits.close()\n",
        "input_fastas.close()\n",
        "output_fasta.close()\n",
        "input_rgs_ids.close()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "id": "9b672d9f",
      "metadata": {
        "id": "9b672d9f"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#017-cat-RGS-CGS-fasta\n",
        "\n",
        "cat source-rgs/rgs5-piezo-full_length-03dec2021.fasta output/16-CGS-final-sequences-by-blastp-RBF.fasta > output/17-AGS_blastp-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "id": "5a3fce1d",
      "metadata": {
        "id": "5a3fce1d"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "#018-sed-dashes-for-mafft-PACBIO\n",
        "\n",
        "# clean sequences for other applications\n",
        "sed -e 's/-/_/g' output/17-AGS_blastp-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa > output/18-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa \n",
        "sed -i 's/U/X/g' output/18-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 55,
      "id": "30f25f0d",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "30f25f0d",
        "outputId": "355847e0-4efb-4613-f8af-51c883668914"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "nthread = 40\n",
            "nthreadpair = 40\n",
            "nthreadtb = 16\n",
            "ppenalty_ex = 0\n",
            "stacksize: 8192 kb\n",
            "rescale = 1\n",
            "Gap Penalty = -1.53, +0.00, +0.00\n",
            "\n",
            "\n",
            "\n",
            "Making a distance matrix ..\n",
            "\r    1 / 5 (thread    9)\n",
            "done.\n",
            "\n",
            "Constructing a UPGMA tree (efffree=0) ... \n",
            "\r    0 / 5\n",
            "done.\n",
            "\n",
            "Progressive alignment 1/2... \n",
            "\rSTEP     1 / 4 (thread    1) f\b\b\rSTEP     2 / 4 (thread   12) f\b\b\rSTEP     3 / 4 (thread   13) f\b\b\n",
            "Reallocating..done. *alloclen = 6735\n",
            "\rSTEP     4 / 4 (thread   14) f\b\b\n",
            "done.\n",
            "\n",
            "Making a distance matrix from msa.. \n",
            "\r    0 / 5 (thread    1)\n",
            "done.\n",
            "\n",
            "Constructing a UPGMA tree (efffree=1) ... \n",
            "\r    0 / 5\n",
            "done.\n",
            "\n",
            "Progressive alignment 2/2... \n",
            "\rSTEP     1 / 4 (thread    1) f\b\b\rSTEP     2 / 4 (thread   14) f\b\b\n",
            "Reallocating..done. *alloclen = 6536\n",
            "\rSTEP     3 / 4 (thread   15) f\b\b\rSTEP     4 / 4 (thread   13) f\b\b\n",
            "done.\n",
            "\n",
            "disttbfast (aa) Version 7.487\n",
            "alg=A, model=BLOSUM45, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
            "16 thread(s)\n",
            "\n",
            "rescale = 1\n",
            "\r    0 / 5 (thread    0)dndpre (aa) Version 7.487\n",
            "alg=X, model=BLOSUM45, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
            "40 thread(s)\n",
            "\n",
            "minimumweight = 0.000010\n",
            "autosubalignment = 0.000000\n",
            "nthread = 8\n",
            "randomseed = 0\n",
            "blosum 45 / kimura 200\n",
            "poffset = 0\n",
            "niter = 16\n",
            "sueff_global = 0.100000\n",
            "nadd = 16\n",
            "rescale = 1\n",
            "\n",
            "\r    0 / 5\n",
            "Segment   1/ 21    1- 608\n",
            "001-0000-0 (thread    3) identical     \r001-0000-0 (thread    3) worse      \r\r001-0001-1 (thread    4) identical     \r001-0001-1 (thread    4) better     \r\r001-0000-1 (thread    5) identical     \r001-0002-1 (thread    5) better     \r\r001-0000-0 (thread    2) identical     \r001-0003-0 (thread    2) better     \r\r001-0004-1 (thread    6) identical     \r001-0004-1 (thread    6) better     \r\r001-0005-1 (thread    7) identical     \r001-0005-1 (thread    7) better     \r\r001-0006-0 (thread    1) identical     \r001-0006-0 (thread    1) worse      \r\r002-0000-1 (thread    6) identical     \r002-0000-1 (thread    6) better     \r\r002-0000-0 (thread    2) identical     \r002-0001-0 (thread    2) worse      \r\r002-0000-1 (thread    4) identical     \r002-0002-1 (thread    4) worse      \r\r002-0003-1 (thread    7) identical     \r002-0003-1 (thread    7) better     \r\r002-0003-1 (thread    3) identical     \r002-0004-1 (thread    3) worse      \r\r002-0003-0 (thread    5) identical     \r002-0005-0 (thread    5) better     \r\r002-0006-0 (thread    8) identical     \r002-0006-0 (thread    8) worse      \r\r003-0000-0 (thread    6) identical     \r003-0000-0 (thread    6) worse      \r\r003-0001-1 (thread    2) identical     \r003-0001-1 (thread    2) better     \r\r003-0001-1 (thread    4) identical     \r003-0002-1 (thread    4) worse      \r\r003-0001-1 (thread    1) identical     \r003-0003-1 (thread    1) worse      \r\r003-0004-1 (thread    7) identical     \r003-0004-1 (thread    7) better     \r\r003-0005-0 (thread    8) identical     \r003-0005-0 (thread    8) worse      \r\r003-0006-0 (thread    3) identical     \r003-0006-0 (thread    3) better     \r\r004-0000-0 (thread    2) identical     \r\r004-0001-1 (thread    6) identical     \r004-0001-1 (thread    6) worse      \r\r004-0002-0 (thread    8) identical     \r004-0002-0 (thread    8) better     \r\r004-0003-1 (thread    4) identical     \r004-0003-1 (thread    4) better     \r\r004-0003-0 (thread    5) identical     \r004-0004-0 (thread    5) worse      \r\r004-0005-1 (thread    1) identical     \r004-0005-1 (thread    1) worse      \r\r004-0006-1 (thread    7) identical     \r004-0006-1 (thread    7) worse      \r\r005-0000-0 (thread    8) identical     \r005-0000-0 (thread    8) worse      \r\r005-0001-1 (thread    4) identical     \r\r005-0002-1 (thread    3) identical     \r005-0002-1 (thread    3) worse      \r\r005-0003-0 (thread    6) identical     \r\r005-0003-0 (thread    5) identical     \r005-0004-0 (thread    5) worse      \r\r005-0005-1 (thread    2) identical     \r005-0005-1 (thread    2) worse      \r\r005-0006-1 (thread    1) identical     \r005-0006-1 (thread    1) worse      \r\r\n",
            "Converged.\n",
            "Segment   2/ 21  589- 846\n",
            "001-0000-0 (thread    2) identical     \r001-0000-0 (thread    2) better     \r\r001-0000-1 (thread    3) identical     \r001-0001-1 (thread    3) better     \r\r001-0002-1 (thread    6) identical     \r001-0002-1 (thread    6) worse      \r\r001-0002-1 (thread    4) identical     \r\r001-0000-1 (thread    8) identical     \r001-0004-1 (thread    8) worse      \r\r001-0000-0 (thread    5) identical     \r001-0005-0 (thread    5) better     \r\r001-0000-0 (thread    7) identical     \r\r002-0000-0 (thread    2) identical     \r002-0000-0 (thread    2) better     \r\r002-0001-0 (thread    8) identical     \r\r002-0002-1 (thread    6) identical     \r002-0002-1 (thread    6) better     \r\r002-0002-1 (thread    8) identical     \r\r002-0004-1 (thread    3) identical     \r002-0004-1 (thread    3) worse      \r\r002-0005-0 (thread    1) identical     \r\r002-0006-1 (thread    2) identical     \r002-0006-1 (thread    2) worse      \r\r003-0000-1 (thread    8) identical     \r003-0000-1 (thread    8) worse      \r\r003-0001-0 (thread    7) identical     \r\r003-0002-1 (thread    3) identical     \r\r003-0002-1 (thread    8) identical     \r003-0003-1 (thread    6) identical     \r003-0003-1 (thread    6) better     \r\r003-0003-1 (thread    8) worse      \r\r003-0005-0 (thread    4) identical     \r\r003-0006-0 (thread    7) identical     \r003-0006-0 (thread    7) worse      \r\r004-0000-0 (thread    5) identical     \r\r004-0001-1 (thread    6) identical     \r004-0001-1 (thread    6) worse      \r\r004-0000-1 (thread    1) identical     \r004-0002-1 (thread    1) worse      \r\r004-0000-0 (thread    8) identical     \r\r004-0004-1 (thread    6) identical     \r004-0004-1 (thread    6) worse      \r\r004-0005-1 (thread    3) identical     \r\r004-0006-0 (thread    5) identical     \r004-0006-0 (thread    5) worse      \r\r\n",
            "Converged.\n",
            "Segment   3/ 21  843-1251\n",
            "001-0000-1 (thread    5) identical     \r001-0000-1 (thread    1) identical     \r001-0000-1 (thread    1) worse      \r\r001-0000-1 (thread    2) identical     \r001-0001-1 (thread    2) worse      \r\r001-0000-1 (thread    6) identical     \r001-0002-1 (thread    6) better     \r\r\r001-0004-0 (thread    8) identical     \r001-0004-0 (thread    8) better     \r\r001-0004-0 (thread    4) identical     \r001-0000-0 (thread    7) identical     \r001-0005-0 (thread    7) worse      \r\r001-0006-0 (thread    4) worse      \r\r002-0000-1 (thread    5) identical     \r002-0000-1 (thread    5) worse      \r\r002-0001-0 (thread    6) identical     \r002-0001-0 (thread    6) worse      \r\r002-0000-1 (thread    2) identical     \r002-0002-1 (thread    2) worse      \r\r002-0003-1 (thread    5) identical     \r002-0003-1 (thread    5) worse      \r\r002-0004-1 (thread    7) identical     \r002-0004-1 (thread    7) better     \r\r002-0005-0 (thread    1) identical     \r002-0005-0 (thread    1) worse      \r\r002-0006-0 (thread    3) identical     \r002-0006-0 (thread    3) worse      \r\r003-0000-0 (thread    6) identical     \r003-0000-0 (thread    6) worse      \r\r003-0001-1 (thread    8) identical     \r003-0001-1 (thread    8) worse      \r\r003-0002-1 (thread    6) identical     \r003-0002-1 (thread    6) worse      \r\r003-0003-0 (thread    2) identical     \r003-0003-0 (thread    2) worse      \r\r003-0004-1 (thread    7) identical     \r003-0004-1 (thread    7) worse      \r\r003-0004-1 (thread    4) identical     \r\r003-0006-0 (thread    5) identical     \r003-0006-0 (thread    5) worse      \r\r\n",
            "Converged.\n",
            "\n",
            "Converged2.\n",
            "Segment   4/ 21 1248-1608\n",
            "001-0000-0 (thread    1) identical     \r\r001-0001-1 (thread    3) identical     \r\r001-0002-1 (thread    2) identical     \r001-0002-1 (thread    2) better     \r\r001-0003-1 (thread    4) identical     \r001-0003-1 (thread    4) worse      \r\r001-0004-0 (thread    5) identical     \r001-0004-0 (thread    5) better     \r\r001-0004-0 (thread    6) identical     \r001-0005-0 (thread    6) better     \r\r001-0004-1 (thread    7) identical     \r001-0006-1 (thread    7) better     \r\r002-0000-1 (thread    8) identical     \r002-0000-1 (thread    8) better     \r\r002-0000-0 (thread    6) identical     \r002-0000-0 (thread    1) identical     \r002-0001-0 (thread    1) better     \r\r\r002-0000-1 (thread    7) identical     \r002-0000-1 (thread    2) identical     \r\r002-0000-1 (thread    3) identical     \r002-0004-1 (thread    3) worse      \r\r002-0005-1 (thread    7) worse      \r\r002-0006-0 (thread    5) identical     \r002-0006-0 (thread    5) better     \r\r003-0000-1 (thread    1) identical     \r003-0000-1 (thread    1) worse      \r\r003-0001-0 (thread    4) identical     \r\r003-0002-0 (thread    4) identical     \r\r003-0002-1 (thread    8) identical     \r003-0003-1 (thread    8) worse      \r\r003-0004-1 (thread    3) identical     \r003-0004-1 (thread    3) worse      \r\r003-0005-1 (thread    2) identical     \r003-0005-1 (thread    2) better     \r\r003-0006-0 (thread    1) identical     \r003-0006-0 (thread    1) better     \r\r004-0000-1 (thread    6) identical     \r004-0000-1 (thread    6) worse      \r\r004-0001-0 (thread    1) identical     \r\r004-0002-0 (thread    6) identical     \r\r004-0002-1 (thread    1) identical     \r004-0003-1 (thread    1) better     \r\r004-0002-1 (thread    8) identical     \r004-0004-1 (thread    8) worse      \r\r004-0002-0 (thread    5) identical     \r004-0005-0 (thread    5) better     \r\r004-0006-1 (thread    7) identical     \r004-0006-1 (thread    7) worse      \r\r005-0000-1 (thread    2) identical     \r005-0000-1 (thread    2) worse      \r\r005-0001-0 (thread    3) identical     \r005-0001-0 (thread    3) better     \r\r005-0002-1 (thread    2) identical     \r005-0002-1 (thread    2) worse      \r\r005-0002-0 (thread    6) identical     \r\r005-0004-1 (thread    3) identical     \r005-0004-1 (thread    3) worse      \r\r005-0005-1 (thread    4) identical     \r005-0005-1 (thread    4) better     \r\r005-0006-0 (thread    1) identical     \r\r006-0000-1 (thread    5) identical     \r006-0000-1 (thread    5) worse      \r\r006-0001-0 (thread    8) identical     \r\r006-0002-1 (thread    8) identical     \r\r006-0003-1 (thread    8) identical     \r006-0003-1 (thread    8) better     \r\r006-0004-0 (thread    5) identical     \r006-0004-0 (thread    5) worse      \r\r006-0004-1 (thread    7) identical     \r006-0005-1 (thread    7) better     \r\r006-0004-0 (thread    2) identical     \r\r007-0000-1 (thread    3) identical     \r007-0000-1 (thread    3) worse      \r\r007-0001-0 (thread    6) identical     \r007-0001-0 (thread    6) better     \r\r007-0002-1 (thread    8) identical     \r\r007-0003-1 (thread    6) identical     \r007-0003-1 (thread    6) worse      \r\r007-0003-1 (thread    1) identical     \r007-0004-1 (thread    1) worse      \r\r007-0005-0 (thread    3) identical     \r007-0005-0 (thread    3) worse      \r\r007-0006-0 (thread    4) identical     \r007-0006-0 (thread    4) worse      \r\r008-0000-1 (thread    5) identical     \r008-0000-1 (thread    5) worse      \r\r008-0000-1 (thread    3) identical     \r008-0001-1 (thread    3) worse      \r\r008-0000-0 (thread    8) identical     \r\r008-0000-0 (thread    7) identical     \r008-0002-0 (thread    7) worse      \r\r008-0000-0 (thread    2) identical     \r008-0003-0 (thread    2) worse      \r\r008-0005-1 (thread    1) identical     \r008-0005-1 (thread    1) worse      \r\r008-0006-1 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "\n",
            "Converged2.\n",
            "Segment   5/ 21 1606-1609\n",
            "001-0000-1 (thread    3) identical     \r\r001-0001-1 (thread    3) identical     \r\r001-0002-1 (thread    3) identical     \r\r001-0003-0 (thread    4) identical     \r\r001-0003-0 (thread    2) identical     \r\r001-0003-0 (thread    6) identical     \r\r001-0003-1 (thread    5) identical     \r\r\n",
            "Converged.\n",
            "Segment   6/ 21 1609-1612\n",
            "001-0000-1 (thread    4) identical     \r001-0000-0 (thread    1) identical     \r\r\r001-0002-1 (thread    3) identical     \r\r001-0003-0 (thread    2) identical     \r\r001-0004-1 (thread    5) identical     \r\r001-0004-1 (thread    8) identical     \r\r001-0004-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "Segment   7/ 21 1612-2259\n",
            "001-0000-1 (thread    3) identical     \r001-0000-1 (thread    3) worse      \r\r001-0001-0 (thread    2) identical     \r001-0001-0 (thread    2) worse      \r\r001-0002-0 (thread    4) identical     \r001-0002-0 (thread    4) better     \r\r001-0002-1 (thread    7) identical     \r\r001-0002-1 (thread    5) identical     \r001-0004-1 (thread    5) worse      \r\r001-0002-0 (thread    6) identical     \r001-0005-0 (thread    6) better     \r\r001-0002-1 (thread    1) identical     \r001-0006-1 (thread    1) better     \r\r002-0000-1 (thread    3) identical     \r002-0000-1 (thread    3) worse      \r\r002-0001-1 (thread    2) identical     \r002-0001-1 (thread    2) worse      \r\r002-0002-1 (thread    1) identical     \r\r002-0003-1 (thread    6) identical     \r002-0003-1 (thread    6) worse      \r\r002-0003-0 (thread    4) identical     \r002-0004-0 (thread    4) worse      \r\r002-0003-0 (thread    8) identical     \r002-0005-0 (thread    8) better     \r\r002-0003-0 (thread    7) identical     \r002-0006-0 (thread    7) better     \r\r003-0000-1 (thread    6) identical     \r003-0000-1 (thread    6) worse      \r\r003-0000-0 (thread    4) identical     \r003-0001-0 (thread    4) better     \r\r003-0002-1 (thread    8) identical     \r003-0002-1 (thread    8) worse      \r\r003-0003-0 (thread    3) identical     \r003-0003-0 (thread    3) worse      \r\r003-0004-1 (thread    1) identical     \r003-0004-1 (thread    1) better     \r\r003-0005-1 (thread    5) identical     \r003-0005-1 (thread    5) worse      \r\r003-0006-0 (thread    2) identical     \r003-0006-0 (thread    2) worse      \r\r004-0000-1 (thread    6) identical     \r004-0000-1 (thread    6) worse      \r\r004-0001-0 (thread    4) identical     \r004-0001-0 (thread    4) worse      \r\r004-0002-1 (thread    8) identical     \r\r004-0002-1 (thread    7) identical     \r004-0003-1 (thread    7) worse      \r004-0002-0 (thread    3) identical     \r004-0003-0 (thread    3) worse      \r\r004-0003-0 (thread    1) identical     \r004-0004-0 (thread    1) worse      \r\r\r004-0006-1 (thread    5) identical     \r004-0006-1 (thread    5) worse      \r\r\n",
            "Converged.\n",
            "Segment   8/ 21 2250-2325\n",
            "001-0000-1 (thread    3) identical     \r001-0000-1 (thread    3) better     \r\r001-0001-0 (thread    2) identical     \r001-0001-0 (thread    2) better     \r\r001-0002-0 (thread    6) identical     \r\r001-0003-1 (thread    7) identical     \r001-0003-1 (thread    7) worse      \r\r001-0000-1 (thread    1) identical     \r\r001-0000-0 (thread    4) identical     \r\r001-0006-1 (thread    5) identical     \r\r002-0000-1 (thread    3) identical     \r\r002-0001-0 (thread    3) identical     \r\r002-0002-1 (thread    3) identical     \r\r002-0003-0 (thread    3) identical     \r\r002-0004-1 (thread    3) identical     \r002-0004-1 (thread    3) worse      \r\r002-0005-1 (thread    3) identical     \r\r002-0006-0 (thread    8) identical     \r\r\n",
            "Converged.\n",
            "Segment   9/ 21 2325-2522\n",
            "001-0000-1 (thread    2) identical     \r001-0000-1 (thread    2) better     \r\r001-0000-1 (thread    1) identical     \r001-0001-1 (thread    1) better     \r\r001-0000-1 (thread    6) identical     \r001-0002-1 (thread    6) worse      \r\r001-0000-0 (thread    8) identical     \r001-0003-0 (thread    8) worse      \r\r001-0004-0 (thread    4) identical     \r001-0004-0 (thread    4) better     \r\r001-0005-1 (thread    5) identical     \r001-0005-1 (thread    5) better     \r\r001-0006-0 (thread    3) identical     \r001-0006-0 (thread    3) worse      \r\r002-0000-1 (thread    2) identical     \r002-0000-1 (thread    2) worse      \r\r002-0000-0 (thread    1) identical     \r\r002-0002-1 (thread    3) identical     \r\r002-0000-0 (thread    4) identical     \r002-0000-1 (thread    6) identical     \r002-0000-1 (thread    8) identical     \r002-0003-1 (thread    8) worse      \r\r002-0003-1 (thread    6) worse      \r\r002-0002-0 (thread    7) identical     \r002-0005-0 (thread    7) worse      \r\r002-0003-0 (thread    4) worse      \r\r\n",
            "Converged.\n",
            "Segment  10/ 21 2522-2526\n",
            "001-0000-1 (thread    7) identical     \r\r001-0001-1 (thread    7) identical     \r\r001-0002-0 (thread    7) identical     \r\r001-0003-1 (thread    7) identical     \r\r001-0004-1 (thread    7) identical     \r\r001-0005-0 (thread    2) identical     \r\r001-0006-0 (thread    1) identical     \r\r\n",
            "Converged.\n",
            "Segment  11/ 21 2526-2570\n",
            "001-0000-1 (thread    1) identical     \r\r001-0001-1 (thread    2) identical     \r\r001-0002-0 (thread    8) identical     \r\r001-0003-1 (thread    4) identical     \r\r001-0001-1 (thread    7) identical     \r\r001-0001-0 (thread    3) identical     \r\r001-0006-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "Segment  12/ 21 2570-2575\n",
            "001-0000-0 (thread    1) identical     \r\r001-0001-1 (thread    1) identical     \r\r001-0002-1 (thread    1) identical     \r\r001-0003-0 (thread    5) identical     \r\r001-0004-1 (thread    2) identical     \r\r001-0005-1 (thread    3) identical     \r\r001-0006-0 (thread    4) identical     \r\r\n",
            "Converged.\n",
            "Segment  13/ 21 2575-2580\n",
            "001-0000-1 (thread    7) identical     \r\r001-0000-0 (thread    3) identical     \r001-0000-1 (thread    8) identical     \r\r001-0000-0 (thread    1) identical     \r\r\r001-0004-0 (thread    2) identical     \r\r001-0005-1 (thread    4) identical     \r\r001-0006-1 (thread    5) identical     \r\r\n",
            "Converged.\n",
            "Segment  14/ 21 2580-2599\n",
            "001-0000-1 (thread    3) identical     \r\r001-0001-0 (thread    4) identical     \r\r001-0002-1 (thread    1) identical     \r\r001-0000-0 (thread    2) identical     \r\r001-0004-0 (thread    6) identical     \r\r001-0004-1 (thread    5) identical     \r\r001-0006-1 (thread    7) identical     \r\r\n",
            "Converged.\n",
            "Segment  15/ 21 2599-2639\n",
            "001-0000-0 (thread    1) identical     \r\r001-0001-0 (thread    1) identical     \r\r001-0002-1 (thread    1) identical     \r\r001-0003-1 (thread    7) identical     \r\r001-0003-1 (thread    5) identical     \r\r001-0003-1 (thread    8) identical     \r\r001-0006-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "Segment  16/ 21 2639-2674\n",
            "001-0000-0 (thread    2) identical     \r\r001-0001-1 (thread    4) identical     \r\r001-0001-1 (thread    5) identical     \r\r001-0003-1 (thread    8) identical     \r\r001-0004-0 (thread    7) identical     \r\r001-0005-0 (thread    3) identical     \r\r001-0006-1 (thread    1) identical     \r\r\n",
            "Converged.\n",
            "Segment  17/ 21 2674-2682\n",
            "001-0000-0 (thread    3) identical     \r\r001-0001-0 (thread    2) identical     \r\r001-0002-1 (thread    4) identical     \r\r001-0003-1 (thread    8) identical     \r\r001-0004-1 (thread    7) identical     \r\r001-0005-1 (thread    1) identical     \r\r001-0006-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "Segment  18/ 21 2682-2687\n",
            "001-0000-0 (thread    1) identical     \r\r001-0001-0 (thread    2) identical     \r\r001-0002-1 (thread    3) identical     \r\r001-0003-1 (thread    7) identical     \r\r001-0003-1 (thread    4) identical     \r\r001-0003-1 (thread    5) identical     \r\r001-0003-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "Segment  19/ 21 2687-2700\n",
            "001-0000-0 (thread    2) identical     \r\r001-0000-1 (thread    8) identical     \r\r001-0002-1 (thread    5) identical     \r\r001-0000-1 (thread    4) identical     \r\r001-0004-0 (thread    7) identical     \r\r001-0005-0 (thread    6) identical     \r\r001-0006-1 (thread    3) identical     \r\r\n",
            "Converged.\n",
            "Segment  20/ 21 2700-3141\n",
            "001-0000-0 (thread    2) identical     \r001-0000-0 (thread    2) worse      \r\r001-0000-0 (thread    5) identical     \r001-0000-1 (thread    4) identical     \r001-0001-1 (thread    4) worse      \r\r001-0000-0 (thread    3) identical     \r001-0002-0 (thread    3) better     \r\r001-0003-0 (thread    5) worse      \r\r001-0004-1 (thread    6) identical     \r001-0004-1 (thread    6) better     \r\r001-0005-1 (thread    7) identical     \r001-0005-1 (thread    7) worse      \r\r001-0005-1 (thread    8) identical     \r001-0006-1 (thread    8) better     \r\r002-0000-1 (thread    5) identical     \r002-0000-1 (thread    5) worse      \r\r002-0001-0 (thread    2) identical     \r002-0001-0 (thread    2) better     \r\r002-0001-1 (thread    3) identical     \r002-0002-1 (thread    3) worse      \r\r002-0001-0 (thread    4) identical     \r002-0003-0 (thread    4) worse      \r\r002-0004-1 (thread    2) identical     \r\r002-0005-0 (thread    1) identical     \r002-0005-0 (thread    1) worse      \r\r002-0006-1 (thread    5) identical     \r002-0006-1 (thread    5) worse      \r\r003-0000-0 (thread    6) identical     \r\r003-0001-0 (thread    4) identical     \r003-0001-0 (thread    4) worse      \r\r003-0001-1 (thread    8) identical     \r003-0001-1 (thread    3) identical     \r003-0001-1 (thread    6) identical     \r003-0002-1 (thread    6) worse      \r\r003-0001-0 (thread    2) identical     \r003-0003-0 (thread    2) worse      \r\r003-0002-1 (thread    8) worse      \r\r003-0002-1 (thread    3) worse      \r\r003-0006-1 (thread    7) identical     \r\r\n",
            "Converged.\n",
            "Segment  21/ 21 3139-3197\n",
            "001-0000-1 (thread    4) identical     \r\r001-0001-0 (thread    1) identical     \r\r001-0002-0 (thread    2) identical     \r\r001-0003-1 (thread    3) identical     \r\r001-0004-1 (thread    5) identical     \r\r001-0005-1 (thread    7) identical     \r\r001-0006-0 (thread    6) identical     \r\r\n",
            "Converged.\n",
            "done\n",
            "dvtditr (aa) Version 7.487\n",
            "alg=A, model=BLOSUM45, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
            "8 thread(s)\n",
            "\n",
            "\n",
            "Strategy:\n",
            " FFT-NS-i (Accurate but slow)\n",
            " Iterative refinement method (max. 16 iterations)\n",
            "\n",
            "If unsure which option to use, try 'mafft --auto input > output'.\n",
            "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
            "\n",
            "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
            "It tends to insert more gaps into gap-rich regions than previous versions.\n",
            "To disable this change, add the --leavegappyregion option.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#019-mafft-pore-sequence-PACBIO\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "mafft --originalseqonly --maxiterate 1000 --reorder  --bl 45 --thread 40 output/18-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa > output/19-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.mafft\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 57,
      "id": "86da8b10",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "86da8b10",
        "outputId": "7808c504-8728-4574-e7f6-c036fb60076d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Determining smart-gap threshold...\n",
            "\n",
            "-------------\n",
            "| Arguments |\n",
            "-------------\n",
            "Input file: output/19-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.mafft (format: fasta)\n",
            "Output file: output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap (format: fasta)\n",
            "Gaps threshold: 0.8\n",
            "Trimming mode: smart-gap\n",
            "Create complementary output: False\n",
            "Create log file: True\n",
            "\n",
            "\n",
            "------------------------\n",
            "| Processing Alignment |\n",
            "------------------------\n",
            "\n",
            "\n",
            "\n",
            "------------------------\n",
            "| Writing output files |\n",
            "------------------------\n",
            "trimmed alignment: output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap\n",
            "complement file: False\n",
            "log file: output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap.log\n",
            "\n",
            "\n",
            "---------------------\n",
            "| Output Statistics |\n",
            "---------------------\n",
            "Number of sites kept: 2734\n",
            "Number of sites trimmed: 462\n",
            "\n",
            "Percentage of alignment trimmed: 14.456%\n",
            "\n",
            "Execution time: 0.179s\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\r  0%|          | 0/3196 [00:00<?, ?it/s]\r 69%|██████▊   | 2192/3196 [00:00<00:00, 21910.93it/s]\r100%|██████████| 3196/3196 [00:00<00:00, 22598.83it/s]\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#020-clipkit-smartgap-pore-sequence_PACBIO\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "clipkit -l -m smart-gap output/19-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.mafft -o output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 60,
      "id": "3a1a17b1",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3a1a17b1",
        "outputId": "63c137aa-b109-44e5-c27e-b65ed65b1f87"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "IQ-TREE multicore version 2.0.3 for Linux 64-bit built Dec 20 2020\n",
            "Developed by Bui Quang Minh, Nguyen Lam Tung, Olga Chernomor,\n",
            "Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
            "\n",
            "Host:    d382aa18806e (AVX2, FMA3, 12 GB RAM)\n",
            "Command: iqtree -s output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap -m MFP --prefix output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED --rate -B 2000 -alrt 2000 -T AUTO -bnni -safe\n",
            "Seed:    900385 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
            "Time:    Fri Feb 18 05:56:01 2022\n",
            "Kernel:  Safe AVX+FMA - auto-detect threads (2 CPU cores detected)\n",
            "\n",
            "Reading alignment file output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap ... Fasta format detected\n",
            "Alignment most likely contains protein sequences\n",
            "Alignment has 5 sequences with 2734 columns, 2295 distinct patterns\n",
            "483 parsimony-informative, 1886 singleton sites, 365 constant sites\n",
            "                                   Gap/Ambiguity  Composition  p-value\n",
            "   1  rgs5_human_PIEZO1_uniprotQ92508      8.89%    failed      0.00%\n",
            "   2  rgs5_human_PIEZO2_uniprotQ9H5I5      5.05%    failed      0.00%\n",
            "   3  rgs5_fly_Piezo_uniprotM9MSG8        10.31%    passed     44.69%\n",
            "   4  rgs5_fly_Pzl_uniprotA0A126GXQ2      16.83%    failed      0.00%\n",
            "   5  rgs5_worm_pezo1_uniprotA0A061ACX2   15.25%    passed     32.72%\n",
            "****  TOTAL                               11.27%  3 sequences failed composition chi2 test (p-value<5%; df=19)\n",
            "\n",
            "\n",
            "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.001 seconds\n",
            "Measuring multi-threading efficiency up to 2 CPU cores\n",
            "Increase to 10 rounds for branch lengths\n",
            "10 trees examined\n",
            "Threads: 1 / Time: 2.162 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -28710\n",
            "Threads: 2 / Time: 1.834 sec / Speedup: 1.179 / Efficiency: 59% / LogL: -28710\n",
            "BEST NUMBER OF THREADS: 2\n",
            "\n",
            "Perform fast likelihood tree search using LG+I+G model...\n",
            "Estimate model parameters (epsilon = 5.000)\n",
            "Perform nearest neighbor interchange...\n",
            "Estimate model parameters (epsilon = 1.000)\n",
            "1. Initial log-likelihood: -27905.263\n",
            "Optimal log-likelihood: -27904.415\n",
            "Proportion of invariable sites: 0.031\n",
            "Gamma shape alpha: 2.345\n",
            "Parameters optimization took 1 rounds (0.085 sec)\n",
            "Time for fast ML tree search: 0.768 seconds\n",
            "\n",
            "NOTE: ModelFinder requires 35 MB RAM!\n",
            "ModelFinder will test up to 546 protein models (sample size: 2734) ...\n",
            " No. Model         -LnL         df  AIC          AICc         BIC\n",
            "  1  LG            28087.037    7   56188.074    56188.115    56229.469\n",
            "  2  LG+I          27995.178    8   56006.357    56006.409    56053.665\n",
            "  3  LG+G4         27901.963    8   55819.926    55819.979    55867.234\n",
            "  4  LG+I+G4       27902.323    9   55822.646    55822.712    55875.868\n",
            "  5  LG+R2         27905.308    9   55828.617    55828.683    55881.838\n",
            "  6  LG+R3         27901.987    11  55825.975    55826.072    55891.024\n",
            " 16  LG+F+G4       27853.695    27  55761.390    55761.949    55921.055\n",
            " 17  LG+F+I+G4     27853.943    28  55763.886    55764.487    55929.465\n",
            " 29  WAG+G4        28085.541    8   56187.081    56187.134    56234.390\n",
            " 30  WAG+I+G4      28084.022    9   56186.044    56186.110    56239.266\n",
            " 42  WAG+F+G4      27914.009    27  55882.019    55882.577    56041.684\n",
            " 43  WAG+F+I+G4    27912.702    28  55881.405    55882.005    56046.983\n",
            " 55  JTT+G4        28100.000    8   56215.999    56216.052    56263.307\n",
            " 56  JTT+I+G4      28099.889    9   56217.778    56217.844    56271.000\n",
            " 68  JTT+F+G4      27969.959    27  55993.918    55994.477    56153.583\n",
            " 69  JTT+F+I+G4    27970.023    28  55996.045    55996.646    56161.624\n",
            " 81  JTTDCMut+G4   28108.230    8   56232.459    56232.512    56279.767\n",
            " 82  JTTDCMut+I+G4 28108.144    9   56234.288    56234.354    56287.510\n",
            " 94  JTTDCMut+F+G4 27973.266    27  56000.533    56001.091    56160.198\n",
            " 95  JTTDCMut+F+I+G4 27973.384    28  56002.767    56003.368    56168.346\n",
            "107  DCMut+G4      28432.116    8   56880.232    56880.285    56927.540\n",
            "108  DCMut+I+G4    28431.863    9   56881.726    56881.792    56934.948\n",
            "120  DCMut+F+G4    28019.359    27  56092.719    56093.277    56252.384\n",
            "121  DCMut+F+I+G4  28019.409    28  56094.819    56095.419    56260.397\n",
            "133  VT+G4         27969.669    8   55955.338    55955.391    56002.646\n",
            "134  VT+I+G4       27967.190    9   55952.379    55952.445    56005.601\n",
            "146  VT+F+G4       27900.967    27  55855.935    55856.493    56015.600\n",
            "147  VT+F+I+G4     27898.710    28  55853.420    55854.020    56018.998\n",
            "159  PMB+G4        28056.541    8   56129.082    56129.135    56176.391\n",
            "160  PMB+I+G4      28053.447    9   56124.893    56124.959    56178.115\n",
            "172  PMB+F+G4      28014.460    27  56082.921    56083.479    56242.586\n",
            "173  PMB+F+I+G4    28011.804    28  56079.609    56080.209    56245.187\n",
            "185  Blosum62+G4   28045.014    8   56106.028    56106.081    56153.336\n",
            "186  Blosum62+I+G4 28042.009    9   56102.018    56102.084    56155.239\n",
            "198  Blosum62+F+G4 27978.038    27  56010.075    56010.634    56169.740\n",
            "199  Blosum62+F+I+G4 27974.927    28  56005.854    56006.455    56171.433\n",
            "211  Dayhoff+G4    28432.303    8   56880.605    56880.658    56927.913\n",
            "212  Dayhoff+I+G4  28432.048    9   56882.097    56882.163    56935.318\n",
            "224  Dayhoff+F+G4  28019.033    27  56092.065    56092.624    56251.730\n",
            "225  Dayhoff+F+I+G4 28019.078    28  56094.155    56094.756    56259.734\n",
            "237  mtREV+G4      29109.846    8   58235.692    58235.745    58283.000\n",
            "238  mtREV+I+G4    29110.605    9   58239.210    58239.276    58292.432\n",
            "250  mtREV+F+G4    28150.552    27  56355.103    56355.662    56514.768\n",
            "251  mtREV+F+I+G4  28151.249    28  56358.499    56359.099    56524.077\n",
            "263  mtART+G4      29190.857    8   58397.714    58397.767    58445.022\n",
            "264  mtART+I+G4    29191.774    9   58401.547    58401.613    58454.769\n",
            "276  mtART+F+G4    28229.241    27  56512.483    56513.042    56672.148\n",
            "277  mtART+F+I+G4  28230.150    28  56516.300    56516.900    56681.878\n",
            "289  mtZOA+G4      28786.849    8   57589.699    57589.751    57637.007\n",
            "290  mtZOA+I+G4    28787.725    9   57593.451    57593.517    57646.672\n",
            "302  mtZOA+F+G4    28023.627    27  56101.254    56101.813    56260.919\n",
            "303  mtZOA+F+I+G4  28024.561    28  56105.122    56105.722    56270.701\n",
            "315  mtMet+G4      29078.758    8   58173.516    58173.569    58220.824\n",
            "316  mtMet+I+G4    29079.662    9   58177.324    58177.390    58230.546\n",
            "328  mtMet+F+G4    28065.523    27  56185.047    56185.605    56344.712\n",
            "329  mtMet+F+I+G4  28066.510    28  56189.020    56189.620    56354.598\n",
            "341  mtVer+G4      29312.436    8   58640.872    58640.925    58688.180\n",
            "342  mtVer+I+G4    29313.310    9   58644.620    58644.686    58697.842\n",
            "354  mtVer+F+G4    28271.632    27  56597.263    56597.822    56756.928\n",
            "355  mtVer+F+I+G4  28272.607    28  56601.215    56601.815    56766.793\n",
            "367  mtInv+G4      29326.098    8   58668.197    58668.250    58715.505\n",
            "368  mtInv+I+G4    29327.079    9   58672.159    58672.225    58725.380\n",
            "380  mtInv+F+G4    27968.763    27  55991.527    55992.086    56151.192\n",
            "381  mtInv+F+I+G4  27969.777    28  55995.554    55996.155    56161.133\n",
            "393  mtMAM+G4      29392.453    8   58800.906    58800.959    58848.215\n",
            "394  mtMAM+I+G4    29393.309    9   58804.618    58804.684    58857.839\n",
            "406  mtMAM+F+G4    28424.033    27  56902.066    56902.625    57061.731\n",
            "407  mtMAM+F+I+G4  28424.802    28  56905.605    56906.205    57071.183\n",
            "419  HIVb+G4       28399.884    8   56815.769    56815.822    56863.077\n",
            "420  HIVb+I+G4     28397.710    9   56813.420    56813.487    56866.642\n",
            "432  HIVb+F+G4     28209.637    27  56473.275    56473.833    56632.940\n",
            "433  HIVb+F+I+G4   28208.836    28  56473.672    56474.272    56639.251\n",
            "445  HIVw+G4       29276.748    8   58569.496    58569.549    58616.804\n",
            "446  HIVw+I+G4     29274.645    9   58567.290    58567.356    58620.511\n",
            "458  HIVw+F+G4     28608.484    27  57270.968    57271.527    57430.633\n",
            "459  HIVw+F+I+G4   28607.248    28  57270.496    57271.097    57436.075\n",
            "471  FLU+G4        28429.132    8   56874.265    56874.318    56921.573\n",
            "472  FLU+I+G4      28426.477    9   56870.954    56871.020    56924.175\n",
            "484  FLU+F+G4      28162.794    27  56379.589    56380.148    56539.254\n",
            "485  FLU+F+I+G4    28160.804    28  56377.608    56378.209    56543.187\n",
            "497  rtREV+G4      28214.517    8   56445.034    56445.087    56492.342\n",
            "498  rtREV+I+G4    28214.116    9   56446.231    56446.297    56499.453\n",
            "510  rtREV+F+G4    27952.684    27  55959.368    55959.927    56119.033\n",
            "511  rtREV+F+I+G4  27952.182    28  55960.364    55960.964    56125.943\n",
            "523  cpREV+G4      28020.164    8   56056.329    56056.382    56103.637\n",
            "524  cpREV+I+G4    28018.705    9   56055.409    56055.475    56108.631\n",
            "536  cpREV+F+G4    27961.686    27  55977.373    55977.931    56137.038\n",
            "537  cpREV+F+I+G4  27960.639    28  55977.278    55977.878    56142.856\n",
            "Akaike Information Criterion:           LG+F+G4\n",
            "Corrected Akaike Information Criterion: LG+F+G4\n",
            "Bayesian Information Criterion:         LG+G4\n",
            "Best-fit model: LG+G4 chosen according to BIC\n",
            "\n",
            "All model information printed to output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.model.gz\n",
            "CPU time for ModelFinder: 121.534 seconds (0h:2m:1s)\n",
            "Wall-clock time for ModelFinder: 63.262 seconds (0h:1m:3s)\n",
            "Generating 2000 samples for ultrafast bootstrap (seed: 900385)...\n",
            "\n",
            "NOTE: 24 MB RAM (0 GB) is required!\n",
            "Estimate model parameters (epsilon = 0.100)\n",
            "1. Initial log-likelihood: -27901.963\n",
            "Optimal log-likelihood: -27901.961\n",
            "Gamma shape alpha: 1.824\n",
            "Parameters optimization took 1 rounds (0.088 sec)\n",
            "Computing ML distances based on estimated model parameters... 0.004 sec\n",
            "Computing BIONJ tree...\n",
            "0.000 seconds\n",
            "Log-likelihood of BIONJ tree: -27901.961\n",
            "--------------------------------------------------------------------\n",
            "|             INITIALIZING CANDIDATE TREE SET                      |\n",
            "--------------------------------------------------------------------\n",
            "Generating 99 parsimony trees... 0.176 second\n",
            "Computing log-likelihood of 4 initial trees ... 0.109 seconds\n",
            "Current best score: -27901.961\n",
            "\n",
            "Do NNI search on 5 best initial trees\n",
            "Finish initializing candidate tree set (5)\n",
            "Current best tree score: -27901.961 / CPU time: 1.114\n",
            "Number of iterations: 5\n",
            "--------------------------------------------------------------------\n",
            "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
            "--------------------------------------------------------------------\n",
            "UPDATE BEST LOG-LIKELIHOOD: -27901.961\n",
            "UPDATE BEST LOG-LIKELIHOOD: -27901.961\n",
            "Iteration 10 / LogL: -27905.138 / Time: 0h:0m:2s\n",
            "Iteration 20 / LogL: -27908.730 / Time: 0h:0m:4s\n",
            "UPDATE BEST LOG-LIKELIHOOD: -27901.961\n",
            "Iteration 30 / LogL: -27902.413 / Time: 0h:0m:6s (0h:0m:16s left)\n",
            "Iteration 40 / LogL: -27906.764 / Time: 0h:0m:9s (0h:0m:14s left)\n",
            "UPDATE BEST LOG-LIKELIHOOD: -27901.961\n",
            "Iteration 50 / LogL: -27902.413 / Time: 0h:0m:11s (0h:0m:11s left)\n",
            "UPDATE BEST LOG-LIKELIHOOD: -27901.961\n",
            "Iteration 60 / LogL: -27901.961 / Time: 0h:0m:13s (0h:0m:9s left)\n",
            "Iteration 70 / LogL: -27904.544 / Time: 0h:0m:15s (0h:0m:6s left)\n",
            "Iteration 80 / LogL: -27905.131 / Time: 0h:0m:18s (0h:0m:4s left)\n",
            "Iteration 90 / LogL: -27901.961 / Time: 0h:0m:21s (0h:0m:2s left)\n",
            "Iteration 100 / LogL: -27906.512 / Time: 0h:0m:24s (0h:0m:0s left)\n",
            "Refining ufboot trees with NNI 1 branch...\n",
            "100 samples done\n",
            "200 samples done\n",
            "300 samples done\n",
            "400 samples done\n",
            "500 samples done\n",
            "600 samples done\n",
            "700 samples done\n",
            "800 samples done\n",
            "900 samples done\n",
            "1000 samples done\n",
            "1100 samples done\n",
            "1200 samples done\n",
            "1300 samples done\n",
            "1400 samples done\n",
            "1500 samples done\n",
            "1600 samples done\n",
            "1700 samples done\n",
            "1800 samples done\n",
            "1900 samples done\n",
            "2000 samples done\n",
            "Total 0 ufboot trees refined\n",
            "TREE SEARCH COMPLETED AFTER 101 ITERATIONS / Time: 0h:2m:22s\n",
            "\n",
            "--------------------------------------------------------------------\n",
            "|                    FINALIZING TREE SEARCH                        |\n",
            "--------------------------------------------------------------------\n",
            "Performs final model parameters optimization\n",
            "Estimate model parameters (epsilon = 0.010)\n",
            "1. Initial log-likelihood: -27901.961\n",
            "Optimal log-likelihood: -27901.961\n",
            "Gamma shape alpha: 1.827\n",
            "Parameters optimization took 1 rounds (0.077 sec)\n",
            "BEST SCORE FOUND : -27901.961\n",
            "Empirical proportions for each category: 0.298 0.187 0.179 0.336\n",
            "Site rates printed to output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.rate\n",
            "\n",
            "Testing tree branches by SH-like aLRT with 2000 replicates...\n",
            "0.264 sec.\n",
            "Creating bootstrap support values...\n",
            "Split supports printed to NEXUS file output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.splits.nex\n",
            "Total tree length: 4.809\n",
            "\n",
            "Total number of iterations: 101\n",
            "CPU time used for tree search: 259.046 sec (0h:4m:19s)\n",
            "Wall-clock time used for tree search: 142.425 sec (0h:2m:22s)\n",
            "Total CPU time used: 260.300 sec (0h:4m:20s)\n",
            "Total wall-clock time used: 143.167 sec (0h:2m:23s)\n",
            "UFBoot trees printed to output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.ufboot\n",
            "\n",
            "Computing bootstrap consensus tree...\n",
            "Reading input file output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.splits.nex...\n",
            "5 taxa and 7 splits.\n",
            "Consensus tree written to output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.contree\n",
            "Reading input trees file output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.contree\n",
            "Log-likelihood of consensus tree: -27901.961\n",
            "\n",
            "Analysis results written to: \n",
            "  IQ-TREE report:                output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.iqtree\n",
            "  Maximum-likelihood tree:       output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.treefile\n",
            "  Likelihood distances:          output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.mldist\n",
            "  Site-specific rates:           output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.rate\n",
            "\n",
            "Ultrafast bootstrap approximation results written to:\n",
            "  Split support values:          output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.splits.nex\n",
            "  Consensus tree:                output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.contree\n",
            "  UFBoot trees:                  output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.ufboot\n",
            "  Screen log file:               output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.log\n",
            "\n",
            "Date and Time: Fri Feb 18 05:59:28 2022\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "#021-iqtree2-PACBIO\n",
        "source $HOME/miniconda3/bin/activate\n",
        "\n",
        "iqtree -s output/20-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.clipkit-smartgap -m MFP --prefix output/21-AGS-rgs5-piezo_X_metazoa16_basal_animals_UPDATED --rate -B 2000 -alrt 2000 -T AUTO -bnni -safe\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 65,
      "id": "181617ae",
      "metadata": {
        "id": "181617ae"
      },
      "outputs": [],
      "source": [
        "#022-USER_INPUT-broken-and-replacement-sequences\n",
        "\n",
        "# #repair_1\n",
        "# Metazoa-Ctenophora-Tentaculata-Lobata-Bolinopsidae-Mnemiopsis-leidyi_PACBIO-Mnemiopsis-leidyi-NonamEVm000133t1\n",
        "# Metazoa-Ctenophora-Tentaculata-Lobata-Bolinopsidae-Mnemiopsis-leidyi_UPDATED-ML018021a-PA\n",
        "# Metazoa-Ctenophora-Tentaculata-Lobata-Bolinopsidae-Mnemiopsis-leidyi_UPDATED-ML018022a-PA\n",
        "# Metazoa-Ctenophora-Tentaculata-Lobata-Bolinopsidae-Mnemiopsis-leidyi_UPDATED-ML018025a-PA\n",
        "# Metazoa-Ctenophora-Tentaculata-Lobata-Bolinopsidae-Mnemiopsis-leidyi_UPDATED-ML018027a-PA"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 66,
      "id": "a6964d1c",
      "metadata": {
        "id": "a6964d1c"
      },
      "outputs": [],
      "source": [
        "#023-python-repair-broken-gene-models-with-pacbio\n",
        "\n",
        "input_repair_ids = open( '022-USER_INPUT-broken-and-replacement-sequences', 'r' )\n",
        "input_ags_broken = open( 'output/17-AGS_blastp-rgs5-piezo_X_metazoa16_basal_animals_UPDATED.aa', 'r' )\n",
        "output_ags = open( 'output/23-AGS_blastp-rgs5-piezo_X_metazoa16_basal_animals_REPAIRED.aa', 'w' )\n",
        "output_dropped = open( 'output/23-dropped-sequences.fasta', 'w' )\n",
        "\n",
        "pacbio_ids = []\n",
        "broken_ids = []\n",
        "count = 0\n",
        "for next_line in input_repair_ids:\n",
        "\n",
        "    if next_line[ 0 ] == '#':\n",
        "        count = 1\n",
        "\n",
        "    elif count == 1:\n",
        "        pacbio_id = next_line[ :-1 ]\n",
        "        pacbio_ids.append( pacbio_id )\n",
        "        count = 0\n",
        "        \n",
        "    else:\n",
        "        broken_id = next_line[ :-1 ]\n",
        "        broken_ids.append( broken_id )\n",
        "\n",
        "ags_initial_seq = {}\n",
        "for next_line in input_ags_broken:\n",
        "\n",
        "    if next_line[ 0 ] == '>':\n",
        "\n",
        "        gene_id = next_line[ 1:-1 ]\n",
        "        ags_initial_seq[ gene_id ] = ''\n",
        "\n",
        "    else:\n",
        "\n",
        "        sequence = next_line[ :-1 ]\n",
        "        ags_initial_seq[ gene_id ] = ags_initial_seq[ gene_id ] + sequence\n",
        "\n",
        "for next_id in ags_initial_seq.keys():\n",
        "\n",
        "    sequence = ags_initial_seq[ next_id ]\n",
        "\n",
        "    if next_id in broken_ids:\n",
        "        output = '>' + next_id + '\\n' + sequence + '\\n'\n",
        "        output_dropped.write( output )\n",
        "\n",
        "    elif next_id in pacbio_ids:\n",
        "        output = '>' + next_id + '\\n' + sequence + '\\n'\n",
        "        output_ags.write( output )\n",
        "\n",
        "    elif len( next_id.split( 'Mnemiopsis' ) ) > 1:\n",
        "\n",
        "        if len( next_id.split( 'UPDATED' ) ) > 1:\n",
        "            output = '>' + next_id + '\\n' + sequence + '\\n'\n",
        "            output_ags.write( output )\n",
        "            \n",
        "        else:\n",
        "            output = '>' + next_id + '\\n' + sequence + '\\n'\n",
        "            output_dropped.write( output )\n",
        "\n",
        "    else:\n",
        "        output = '>' + next_id + '\\n' + sequence + '\\n'\n",
        "        output_ags.write( output )\n",
        "\n",
        "input_repair_ids.close()\n",
        "input_ags_broken.close()\n",
        "output_ags.close()\n",
        "output_dropped.close()\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.3"
    },
    "colab": {
      "name": "6-homologs-genes.ipynb",
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}